--- net-mgmt/fprobe/files/fprobe.in.orig	2012-12-18 13:50:22.000000000 +0100
+++ net-mgmt/fprobe/files/fprobe.in	2012-12-18 13:51:43.000000000 +0100
@@ -17,9 +17,11 @@
 
 name="%%NAME%%"
 rcvar=%%NAME%%_enable
+load_rc_config $name
+: ${fprope_enable:=no}
+: ${fprobe_flags:="-i lo0 127.0.0.1:2055"}
 
 command="%%PREFIX%%/sbin/fprobe"
-command_args="-i lo0 127.0.0.1:2055"
 pidfile="/var/run/$name.pid"
 
 load_rc_config "$name"
