Index: tools/tools/netmap/pkt-gen.c
===================================================================
--- tools/tools/netmap/pkt-gen.c	(revision 307401)
+++ tools/tools/netmap/pkt-gen.c	(working copy)
@@ -241,6 +241,7 @@
 	int system_cpus;	/* cpus on the system */
 
 	int options;	/* testing */
+	int softchecksum;	/* Software UDP checksum calculation */
 #define OPT_PREFETCH	1
 #define OPT_ACCESS	2
 #define OPT_COPY	4
@@ -647,6 +648,7 @@
 {
 	uint32_t a;
 	uint16_t p;
+	struct ether_header *eh = &pkt->eh;
 	struct ip *ip = &pkt->ip;
 	struct udphdr *udp = &pkt->udp;
 
@@ -693,6 +695,25 @@
 	ip->ip_dst.s_addr = htonl(g->dst_ip.start);
     } while (0);
     // update checksum
+	/* Some NIC disable hardware CRC checksum in netmap mode
+	 * and don't allow to re-enable it */
+
+	if (g->softchecksum) {
+		ip->ip_sum = 0;
+		ip->ip_sum = wrapsum(checksum(ip, sizeof(*ip), 0));
+		/* UDP checksum */
+		uint16_t paylen = g->pkt_size - sizeof(*eh) - sizeof(struct ip);
+		udp->uh_sum = 0;
+		udp->uh_sum = wrapsum(checksum(udp, sizeof(*udp),
+			checksum(pkt->body,
+				paylen - sizeof(*udp),
+				checksum(&ip->ip_src, 2 * sizeof(ip->ip_src),
+					IPPROTO_UDP + (u_int32_t)ntohs(udp->uh_ulen)
+				)
+			)
+		));
+	}
+
 }
 
 /*
@@ -1891,6 +1912,7 @@
 		"\t-E pipes		allocate extra space for a number of pipes\n"
 		"\t-r			do not touch the buffers (send rubbish)\n"
 	        "\t-P file		load packet from pcap file\n"
+		"\t-U			enable software UDP CRC chekcsum\n"
 		"\t-z			use random IPv4 src address/port\n"
 		"\t-Z			use random IPv4 dst address/port\n"
 		"\t-F num_frags		send multi-slot packets\n"
@@ -2328,6 +2350,9 @@
 		case 'b':	/* burst */
 			g.burst = atoi(optarg);
 			break;
+		case 'U':	/* software UDP checksum */
+			g.softchecksum = 1;
+			break;
 		case 'c':
 			g.cpus = atoi(optarg);
 			break;
