Index: sys/netpfil/pf/pf_norm.c
===================================================================
--- sys/netpfil/pf/pf_norm.c	(revision 300298)
+++ sys/netpfil/pf/pf_norm.c	(working copy)
@@ -374,7 +374,7 @@
 		}
 
 		*(struct pf_fragment_cmp *)frag = *key;
-		frag->fr_timeout = time_second;
+		frag->fr_timeout = time_uptime;
 		frag->fr_maxlen = frent->fe_len;
 		TAILQ_INIT(&frag->fr_queue);
 
