Index: sbin/init/rc.conf
===================================================================
--- sbin/init/rc.conf	(revision 329245)
+++ sbin/init/rc.conf	(working copy)
@@ -599,7 +599,7 @@
 chkprintcap_enable="NO"	# Run chkprintcap(8) before running lpd.
 chkprintcap_flags="-d"	# Create missing directories by default.
 dumpdev="AUTO"		# Device to crashdump to (device name, AUTO, or NO).
-dumpon_flags=""		# Options to pass to dumpon(8), followed by dumpdev.
+dumpon_flags="-Z"		# Options to pass to dumpon(8), followed by dumpdev.
 dumpdir="/var/crash"	# Directory where crash dumps are to be stored
 savecore_enable="YES"	# Extract core from dump devices if any
 savecore_flags="-m 10"	# Used if dumpdev is enabled above, and present.
