--- nanobsd.sh.orig	2011-08-30 21:11:52.000000000 +0200
+++ nanobsd.sh	2011-08-30 21:12:44.000000000 +0200
@@ -533,7 +533,11 @@
 	populate_slice /dev/${MD}s1a ${NANO_WORLDDIR} ${MNT} "s1a"
 	mount /dev/${MD}s1a ${MNT}
 	echo "Generating mtree..."
-	( cd ${MNT} && mtree -c ) > ${NANO_OBJ}/_.mtree
+	echo "./etc" > ${NANO_OBJ}/mtree-exclude
+    echo "./var" >> ${NANO_OBJ}/mtree-exclude
+    echo "./tmp" >> ${NANO_OBJ}/mtree-exclude
+    echo "./dev" >> ${NANO_OBJ}/mtree-exclude
+	( cd ${MNT} && mtree -x -ic -k flags,gid,mode,nlink,size,link,uid,sha256digest -X ${NANO_OBJ}/mtree-exclude ) > ${NANO_OBJ}/_.mtree
 	( cd ${MNT} && du -k ) > ${NANO_OBJ}/_.du
 	umount ${MNT}
 
