--- nanobsd.sh.orig	2010-12-31 15:38:56.000000000 +0100
+++ nanobsd.sh	2010-12-31 15:48:14.000000000 +0100
@@ -135,6 +135,9 @@
 # Can be "file" or "swap"
 NANO_MD_BACKING="file"
 
+# for swap type md(4) backing, write out the mbr only
+NANO_IMAGE_MBRONLY=true
+
 # Progress Print level
 PPLEVEL=3
 
@@ -166,6 +169,13 @@
 #
 #######################################################################
 
+nano_cleanup() {
+   if [ $? -ne 0 ]; then
+       echo "Error encountered.  Check for errors in last log file." 1>&2
+    fi
+    exit $?
+}
+
 clean_build ( ) (
 	pprint 2 "Clean and create object directory (${MAKEOBJDIRPREFIX})"
 
@@ -390,7 +400,7 @@
 prune_usr() (
 
 	# Remove all empty directories in /usr 
-	find ${NANO_WORLDDIR}/usr -type d -depth -print |
+	find ${NANO_WORLDDIR}/usr -type d -depth -not -name aout -print |
 		while read d
 		do
 			rmdir $d > /dev/null 2>&1 || true 
@@ -563,8 +573,13 @@
 	fi
 
 	if [ "${NANO_MD_BACKING}" = "swap" ] ; then
-		echo "Writing out ${NANO_IMGNAME}..."
-		dd if=/dev/${MD} of=${IMG} bs=64k
+       if [ ${NANO_IMAGE_MBRONLY} ]; then
+           echo "Writing out _.disk.mbr..."
+           dd if=/dev/${MD} of=${NANO_DISKIMGDIR}/_.disk.mbr bs=512 count=1
+       else
+            echo "Writing out ${NANO_IMGNAME}..."
+            dd if=/dev/${MD} of=${IMG} bs=64k
+        fi
 	fi
 
 	echo "Writing out _.disk.image..."
@@ -572,6 +587,7 @@
 	mdconfig -d -u $MD
 
 	trap - 1 2 15 EXIT
+	trap "nano_cleanup" EXIT
 
 	) > ${NANO_OBJ}/_.di 2>&1
 )
@@ -658,7 +674,7 @@
 	sed -i "" -e '/^ttyv[0-8]/s/	on/	off/' ${NANO_WORLDDIR}/etc/ttys
 
 	# Tell loader to use serial console early.
-	echo " -h" > ${NANO_WORLDDIR}/boot.config
+	echo ${NANO_BOOT2CFG} > ${NANO_WORLDDIR}/boot.config
 )
 
 #######################################################################
@@ -763,7 +779,7 @@
 	echo "	-i	suppress disk image build"
 	echo "	-k	suppress buildkernel"
 	echo "	-n	add -DNO_CLEAN to buildworld, buildkernel, etc"
-	echo "	-q	make output more quiet"
+	echo "	-q	make output quieter"
 	echo "	-v	make output more verbose"
 	echo "	-w	suppress buildworld"
 	echo "	-c	specify config file"
@@ -839,6 +855,7 @@
 	echo "$0: Extraneous arguments supplied"
 	usage
 fi
+trap "nano_cleanup" EXIT
 
 #######################################################################
 # Setup and Export Internal variables
