Index: sys/dev/ixgbe/if_ix.c
===================================================================
--- sys/dev/ixgbe/if_ix.c	(revision 320907)
+++ sys/dev/ixgbe/if_ix.c	(working copy)
@@ -728,6 +728,8 @@
 	ctrl_ext |= IXGBE_CTRL_EXT_DRV_LOAD;
 	IXGBE_WRITE_REG(hw, IXGBE_CTRL_EXT, ctrl_ext);
 
+	hw->allow_unsupported_sfp = allow_unsupported_sfp;
+
 	/*
 	 * Initialize the shared code
 	 */
@@ -740,8 +742,6 @@
 	if (hw->mbx.ops.init_params)
 		hw->mbx.ops.init_params(hw);
 
-	hw->allow_unsupported_sfp = allow_unsupported_sfp;
-
 	/* Pick up the 82599 settings */
 	if (hw->mac.type != ixgbe_mac_82598EB) {
 		hw->phy.smart_speed = ixgbe_smart_speed;
