diff --git a/sys/dev/if_wg/module/wg_noise.c b/sys/dev/if_wg/module/wg_noise.c
index 946a570916a..5a438151e70 100644
--- a/sys/dev/if_wg/module/wg_noise.c
+++ b/sys/dev/if_wg/module/wg_noise.c
@@ -108,7 +108,7 @@ noise_local_set_private(struct noise_local *l, uint8_t private[NOISE_KEY_SIZE])
 
 	memcpy(l->l_private, private, NOISE_KEY_SIZE);
 	curve25519_clamp_secret(l->l_private);
-	l->l_has_identity = curve25519_generate_public(l->l_public, private);
+	l->l_has_identity = curve25519_generate_public(l->l_public, l->l_private);
 
 	return l->l_has_identity ? 0 : ENXIO;
 }
