Index: sys/dev/mlx5/mlx5_en/mlx5_en_tx.c
===================================================================
--- sys/dev/mlx5/mlx5_en/mlx5_en_tx.c	(revision 354403)
+++ sys/dev/mlx5/mlx5_en/mlx5_en_tx.c	(working copy)
@@ -609,7 +609,7 @@
 	struct mlx5e_sq *sq;
 	int ret;
 
-	if (mb->m_pkthdr.snd_tag != NULL) {
+	if (0 && mb->m_pkthdr.snd_tag != NULL) {
 		sq = mlx5e_select_queue_by_send_tag(ifp, mb);
 		if (unlikely(sq == NULL)) {
 			/* Check for route change */
