Index: tools/tools/ifinfo/Makefile
===================================================================
--- tools/tools/ifinfo/Makefile	(revision 356193)
+++ tools/tools/ifinfo/Makefile	(working copy)
@@ -3,5 +3,6 @@
 PROG=	ifinfo
 SRCS=	ifinfo.c rfc1650.c
 MAN=
+BINDIR=	/usr/local/bin
 
 .include <bsd.prog.mk>
Index: tools/tools/mctest/Makefile
===================================================================
--- tools/tools/mctest/Makefile	(revision 356193)
+++ tools/tools/mctest/Makefile	(working copy)
@@ -2,5 +2,6 @@
 
 PROG_CXX=   mctest 
 LIBADD= pthread
+BINDIR=	/usr/local/bin
 
 .include <bsd.prog.mk>
Index: tools/tools/netrate/netsend/Makefile
===================================================================
--- tools/tools/netrate/netsend/Makefile	(revision 356193)
+++ tools/tools/netrate/netsend/Makefile	(working copy)
@@ -5,5 +5,6 @@
 CFLAGS+=	-Wall
 PROG=	netsend
 MAN=
+BINDIR?=	/usr/local/bin
 
 .include <bsd.prog.mk>
