diff --git usr.sbin/pkg/pkg.c usr.sbin/pkg/pkg.c
index 6138ca37e43..06b0d88e4a7 100644
--- usr.sbin/pkg/pkg.c
+++ usr.sbin/pkg/pkg.c
@@ -843,9 +843,9 @@ bootstrap_pkg(bool force, const char *fetchOpts)
 			if (verify_pubsignature(fd_pkg, fd_sig) == false)
 				goto cleanup;
 		} else {
-			warnx("Signature type %s is not supported for "
+			warnx("Signature type %s is officially not supported for "
 			    "bootstrapping.", signature_type);
-			goto cleanup;
+			/* goto cleanup; */
 		}
 	}
 
@@ -970,7 +970,7 @@ bootstrap_pkg_local(const char *pkgpath, bool force)
 		} else {
 			warnx("Signature type %s is not supported for "
 			    "bootstrapping.", signature_type);
-			goto cleanup;
+			/* goto cleanup; */
 		}
 	}
 
