Index: libexec/rc/rc.d/random
===================================================================
--- libexec/rc/rc.d/random	(revision 346264)
+++ libexec/rc/rc.d/random	(working copy)
@@ -83,7 +83,17 @@
 	[Nn][Oo])
 		;;
 	*)
-		save_dev_random "${entropy_boot_file}"
+		if [ -f ${entropy_boot_file} ]; then
+			if [ $(sha256 -q ${entropy_boot_file}) = "9f5de1687c7c9069da2584b19388aa41d0b5725b5e1530658d815824b261075a" ]; then
+				echo "Factory default ${entropy_boot_file} file detected, replacing it..."
+				mount -uw /
+				save_dev_random "${entropy_boot_file}"
+				mount -ur /
+			fi
+		else
+				save_dev_random "${entropy_boot_file}"
+
+		fi
 		;;
 	esac
 
