Index: usr.bin/netstat/if.c
===================================================================
--- usr.bin/netstat/if.c	(revision 284539)
+++ usr.bin/netstat/if.c	(working copy)
@@ -165,7 +165,7 @@
  * Display a formatted value, or a '-' in the same space.
  */
 static void
-show_stat(const char *fmt, int width, u_long value, short showvalue)
+show_stat(const char *fmt, int width, u_long value, short showvalue, short notbyte)
 {
 	const char *lsep, *rsep;
 	char newfmt[32];
@@ -191,8 +191,14 @@
 		char buf[5];
 
 		/* Format in human readable form. */
-		humanize_number(buf, sizeof(buf), (int64_t)value, "",
-		    HN_AUTOSCALE, HN_NOSPACE | HN_DECIMAL);
+		if (notbyte == 0) {
+			humanize_number(buf, sizeof(buf), (int64_t)value, "",
+				HN_AUTOSCALE, HN_NOSPACE | HN_DECIMAL);
+		} else {
+			humanize_number(buf, sizeof(buf), (int64_t)value, "",
+				HN_AUTOSCALE, HN_NOSPACE | HN_DECIMAL |
+				HN_DIVISOR_1000);
+		}
 		sprintf(newfmt, "%s%%%ds%s", lsep, width, rsep);
 		printf(newfmt, buf);
 	} else {
@@ -287,7 +293,7 @@
 			printf("%-5.5s", ifa->ifa_name);
 
 #define IFA_MTU(ifa)	(((struct if_data *)(ifa)->ifa_data)->ifi_mtu)
-		show_stat("lu", 6, IFA_MTU(ifa), IFA_MTU(ifa));
+		show_stat("lu", 6, IFA_MTU(ifa), IFA_MTU(ifa), 1);
 #undef IFA_MTU
 
 		switch (ifa->ifa_addr->sa_family) {
@@ -373,18 +379,18 @@
 		}
 
 #define	IFA_STAT(s)	(((struct if_data *)ifa->ifa_data)->ifi_ ## s)
-		show_stat("lu", 8, IFA_STAT(ipackets), link|network);
-		show_stat("lu", 5, IFA_STAT(ierrors), link);
-		show_stat("lu", 5, IFA_STAT(iqdrops), link);
+		show_stat("lu", 8, IFA_STAT(ipackets), link|network, 1);
+		show_stat("lu", 5, IFA_STAT(ierrors), link, 1);
+		show_stat("lu", 5, IFA_STAT(iqdrops), link, 1);
 		if (bflag)
-			show_stat("lu", 10, IFA_STAT(ibytes), link|network);
-		show_stat("lu", 8, IFA_STAT(opackets), link|network);
-		show_stat("lu", 5, IFA_STAT(oerrors), link);
+			show_stat("lu", 10, IFA_STAT(ibytes), link|network, 0);
+		show_stat("lu", 8, IFA_STAT(opackets), link|network, 1);
+		show_stat("lu", 5, IFA_STAT(oerrors), link, 1);
 		if (bflag)
-			show_stat("lu", 10, IFA_STAT(obytes), link|network);
-		show_stat("NRSlu", 5, IFA_STAT(collisions), link);
+			show_stat("lu", 10, IFA_STAT(obytes), link|network, 0);
+		show_stat("NRSlu", 5, IFA_STAT(collisions), link, 1);
 		if (dflag)
-			show_stat("LSlu", 5, IFA_STAT(oqdrops), link);
+			show_stat("LSlu", 5, IFA_STAT(oqdrops), link, 1);
 		putchar('\n');
 
 		if (!aflag)
@@ -569,16 +575,16 @@
 
 	fill_iftot(new);
 
-	show_stat("lu", 10, new->ift_ip - old->ift_ip, 1);
-	show_stat("lu", 5, new->ift_ie - old->ift_ie, 1);
-	show_stat("lu", 5, new->ift_id - old->ift_id, 1);
-	show_stat("lu", 10, new->ift_ib - old->ift_ib, 1);
-	show_stat("lu", 10, new->ift_op - old->ift_op, 1);
-	show_stat("lu", 5, new->ift_oe - old->ift_oe, 1);
-	show_stat("lu", 10, new->ift_ob - old->ift_ob, 1);
-	show_stat("NRSlu", 5, new->ift_co - old->ift_co, 1);
+	show_stat("lu", 10, new->ift_ip - old->ift_ip, 1, 1);
+	show_stat("lu", 5, new->ift_ie - old->ift_ie, 1, 1);
+	show_stat("lu", 5, new->ift_id - old->ift_id, 1, 1);
+	show_stat("lu", 10, new->ift_ib - old->ift_ib, 1, 0);
+	show_stat("lu", 10, new->ift_op - old->ift_op, 1, 1);
+	show_stat("lu", 5, new->ift_oe - old->ift_oe, 1, 1);
+	show_stat("lu", 10, new->ift_ob - old->ift_ob, 1, 0);
+	show_stat("NRSlu", 5, new->ift_co - old->ift_co, 1, 1);
 	if (dflag)
-		show_stat("LSlu", 5, new->ift_od - old->ift_od, 1);
+		show_stat("LSlu", 5, new->ift_od - old->ift_od, 1, 1);
 	putchar('\n');
 	fflush(stdout);
 
