Index: tools/tools/ifinfo/Makefile
===================================================================
--- tools/tools/ifinfo/Makefile	(revision 315217)
+++ tools/tools/ifinfo/Makefile	(working copy)
@@ -3,5 +3,6 @@
 PROG=	ifinfo
 SRCS=	ifinfo.c rfc1650.c
 MAN=
+BINDIR=	/usr/local/bin
 
 .include <bsd.prog.mk>
Index: tools/tools/mctest/Makefile
===================================================================
--- tools/tools/mctest/Makefile	(revision 315217)
+++ tools/tools/mctest/Makefile	(working copy)
@@ -2,5 +2,6 @@
 
 PROG_CXX=   mctest 
 LIBADD= pthread
+BINDIR=	/usr/local/bin
 
 .include <bsd.prog.mk>
Index: tools/tools/netrate/netsend/Makefile
===================================================================
--- tools/tools/netrate/netsend/Makefile	(revision 315217)
+++ tools/tools/netrate/netsend/Makefile	(working copy)
@@ -5,5 +5,6 @@
 CFLAGS+=	-Wall
 PROG=	netsend
 MAN=
+BINDIR?=	/usr/local/bin
 
 .include <bsd.prog.mk>
Index: tools/tools/netmap/Makefile
===================================================================
--- tools/tools/netmap/Makefile	(revision 354229)
+++ tools/tools/netmap/Makefile	(working copy)
@@ -4,6 +4,7 @@
 # For multiple programs using a single source file each,
 # we can just define 'progs' and create custom targets.
 PROGS	=	pkt-gen nmreplay bridge lb
+BINDIR	=	/usr/local/sbin
 
 CLEANFILES = $(PROGS) *.o
 MAN=
@@ -34,3 +35,8 @@
 
 lb: lb.o pkt_hash.o
 	$(CC) $(CFLAGS) -o lb lb.o pkt_hash.o $(LDFLAGS)
+
+install: ${PROGS}
+	for i in ${PROGS}; do \
+		install $$i ${DESTDIR}${BINDIR}; \
+	done
