Index: sys/dev/re/if_re.c
===================================================================
--- sys/dev/re/if_re.c	(revision 255043)
+++ sys/dev/re/if_re.c	(working copy)
@@ -233,6 +233,7 @@
 	{ RL_HWREV_8168E, RL_8169, "8168E/8111E", RL_JUMBO_MTU_9K},
 	{ RL_HWREV_8168E_VL, RL_8169, "8168E/8111E-VL", RL_JUMBO_MTU_6K},
 	{ RL_HWREV_8168F, RL_8169, "8168F/8111F", RL_JUMBO_MTU_9K},
+        { RL_HWREV_8168G, RL_8169, "8168G/8111G", RL_JUMBO_MTU_9K},
 	{ RL_HWREV_8411, RL_8169, "8411", RL_JUMBO_MTU_9K},
 	{ 0, 0, NULL, 0 }
 };
@@ -1456,6 +1457,7 @@
 		break;
 	case RL_HWREV_8168E_VL:
 	case RL_HWREV_8168F:
+        case RL_HWREV_8168G:
 	case RL_HWREV_8411:
 		sc->rl_flags |= RL_FLAG_PHYWAKE | RL_FLAG_PAR |
 		    RL_FLAG_DESCV2 | RL_FLAG_MACSTAT | RL_FLAG_CMDSTOP |
Index: sys/pci/if_rlreg.h
===================================================================
--- sys/pci/if_rlreg.h	(revision 255043)
+++ sys/pci/if_rlreg.h	(working copy)
@@ -190,6 +190,7 @@
 #define	RL_HWREV_8105E_SPIN1	0x40C00000
 #define	RL_HWREV_8402		0x44000000
 #define	RL_HWREV_8168F		0x48000000
+#define RL_HWREV_8168G		0x4c000000
 #define	RL_HWREV_8411		0x48800000
 #define	RL_HWREV_8139		0x60000000
 #define	RL_HWREV_8139A		0x70000000
