Index: sys/netpfil/pf/pf_norm.c
===================================================================
--- sys/netpfil/pf/pf_norm.c	(revision 300297)
+++ sys/netpfil/pf/pf_norm.c	(working copy)
@@ -434,7 +434,7 @@
 
 		*(struct pf_fragment_cmp *)frag = *key;
 		frag->fr_flags = 0;
-		frag->fr_timeout = time_second;
+		frag->fr_timeout = time_uptime;
 		frag->fr_maxlen = frent->fe_len;
 		TAILQ_INIT(&frag->fr_queue);
 
