Index: tools/build/mk/OptionalObsoleteFiles.inc
===================================================================
--- tools/build/mk/OptionalObsoleteFiles.inc	(revision 343792)
+++ tools/build/mk/OptionalObsoleteFiles.inc	(working copy)
@@ -6280,7 +6280,6 @@
 .if ${MK_NLS} == no
 OLD_DIRS+=usr/share/nls/
 OLD_DIRS+=usr/share/nls/C
-OLD_FILES+=usr/share/mk/bsd.nls.mk
 OLD_FILES+=usr/share/nls/C/ee.cat
 OLD_DIRS+=usr/share/nls/af_ZA.ISO8859-1
 OLD_DIRS+=usr/share/nls/af_ZA.ISO8859-15
