Index: sbin/dhclient/bpf.c
===================================================================
--- sbin/dhclient/bpf.c	(revision 335629)
+++ sbin/dhclient/bpf.c	(working copy)
@@ -175,6 +175,8 @@
 	if (setsockopt(sock, IPPROTO_IP, IP_HDRINCL, &on,
 	    sizeof(on)) == -1)
 		error("setsockopt(IP_HDRINCL): %m");
+	if (setsockopt(sock, SOL_SOCKET, SO_SETFIB, &info->fib, sizeof(info->fib)))
+		error("setsockopt(SOL_SOCKET, SO_SETFIB): %m");
 	info->ufdesc = sock;
 }
 
Index: sbin/dhclient/dhclient.c
===================================================================
--- sbin/dhclient/dhclient.c	(revision 335629)
+++ sbin/dhclient/dhclient.c	(working copy)
@@ -374,6 +374,7 @@
 	int			 ch, fd, quiet = 0, i = 0;
 	int			 pipe_fd[2];
 	int			 immediate_daemon = 0;
+	uint32_t	 fib = 0;
 	struct passwd		*pw;
 	pid_t			 otherpid;
 	cap_rights_t		 rights;
@@ -395,6 +396,9 @@
 		case 'd':
 			no_daemon = 1;
 			break;
+		case 'f':
+			fib = atoi(optarg);
+			break;
 		case 'l':
 			path_dhclient_db = optarg;
 			break;
@@ -443,6 +447,7 @@
 	if (quiet)
 		log_perror = 0;
 
+	ifi->fib = fib;
 	tzset();
 	time(&cur_time);
 
@@ -563,8 +568,8 @@
 usage(void)
 {
 
-	fprintf(stderr, "usage: %s [-bdqu] ", getprogname());
-	fprintf(stderr, "[-c conffile] [-l leasefile] interface\n");
+	fprintf(stderr, "usage: %s [-bdqu] ", getprogname());
+	fprintf(stderr, "[-c conffile] [-l leasefile] [-f fib] interface\n");
 	exit(1);
 }
 
Index: sbin/dhclient/dhcpd.h
===================================================================
--- sbin/dhclient/dhcpd.h	(revision 335629)
+++ sbin/dhclient/dhcpd.h	(working copy)
@@ -215,6 +215,7 @@
 	int			 dead;
 	u_int16_t		 index;
 	int			 linkstat;
+	uint32_t	 fib;
 };
 
 struct timeout {
