--- etc/rc.initdiskless	2017/06/15 19:56:59	319986
+++ etc/rc.initdiskless	2017/06/15 20:06:41	319987
@@ -195,10 +195,10 @@
     to_umount="$b ${to_umount}"
 }
 
-# Create a generic memory disk
+# Create a generic memory disk (using tmpfs)
 #
 mount_md() {
-    /sbin/mdmfs -S -i 4096 -s $1 -M md $2
+    mount -t tmpfs -o size=$(($1 * 512)) tmpfs $2
 }
 
 # Create the memory filesystem if it has not already been created
