Index: net/bird/Makefile
===================================================================
--- net/bird/Makefile	(revision 465813)
+++ net/bird/Makefile	(working copy)
@@ -1,32 +1,56 @@
-# Created by: Pav Lucistnik <pav@FreeBSD.org>
+# Created by: Alexander V. Chernikov <melifaro@FreeBSD.org>
 # $FreeBSD$
 
 PORTNAME=	bird
-PORTVERSION=	1.6.4
+DISTVERSION=	2.0.2
 CATEGORIES=	net
 MASTER_SITES=	ftp://bird.network.cz/pub/bird/
 
 MAINTAINER=	olivier@FreeBSD.org
-COMMENT?=	Dynamic IP routing daemon (${FLAVOR:Uipv4:S/ip/IP/} version)
+COMMENT=	Dynamic IP routing daemon
 
 LICENSE=	GPLv2
 
-FLAVORS=	ipv4 ipv6
-ipv6_PKGNAMESUFFIX=	6
+LIB_DEPENDS=	libssh.so:security/libssh
 
-USES=		autoreconf bison gmake ncurses readline
+CONFLICTS=	bird-[0-9]*
+CONFLICTS+=	bird6-[0-9]*
+
+USES=		bison gmake ncurses readline
 GNU_CONFIGURE=	yes
 CONFIGURE_ARGS=	--localstatedir=/var
+USE_RC_SUBR=	bird
+SUB_FILES=	pkg-message
 
 MAKE_JOBS_UNSAFE=	yes
 
-USE_RC_SUBR=	${PKGBASE}
+OPTIONS_MULTI=	RP
+RP_DESC=	Routing Protocols
+OPTIONS_MULTI_RP=	BFD BABEL BGP OSPF PIPE RADV RIP STATIC
+OPTIONS_DEFAULT=	${OPTIONS_MULTI_RP}
 
-.if ${FLAVOR:U} == ipv6
-CONFIGURE_ARGS+=	--enable-ipv6
-PLIST_SUB=	VER=6
-.else
-PLIST_SUB?=	VER=""
-.endif
+BFD_DESC=	Bidirectional Forwarding Detection
+BABEL_DESC=	Babel routing protocol
+BGP_DESC=	Border Gateway Protocol
+OSPF_DESC=	Open Short Path First
+PIPE_DESC=	PIPE routing
+RADV_DESC=	Router Advertisement
+RIP_DESC=	Routing Information Protocol
+STATIC_DESC=	Static routing
 
+BFD_VARS=	rt_prot+=bfd
+BABEL_VARS=	rt_prot+=babel
+BGP_VARS=	rt_prot+=bgp
+OSPF_VARS=	rt_prot+=ospf
+PIPE_VARS=	rt_prot+=pipe
+RADV_VARS=	rt_prot+=radv
+RIP_VARS=	rt_prot+=rip
+STATIC_VARS=	rt_prot+=static
+
+CONFIGURE_ARGS+=--with-protocols="${RT_PROT}"
+
+post-patch:
+	${REINPLACE_CMD} -e 's|\.example|\.sample|g' ${WRKSRC}/Makefile.in
+	${MV} ${WRKSRC}/doc/bird.conf.example ${WRKSRC}/doc/bird.conf.sample
+
 .include <bsd.port.mk>
Index: net/bird/distinfo
===================================================================
--- net/bird/distinfo	(revision 465813)
+++ net/bird/distinfo	(working copy)
@@ -1,3 +1,3 @@
-TIMESTAMP = 1522168475
-SHA256 (bird-1.6.4.tar.gz) = c26b8caae988dba81a9dbbee93502463d4326d1b749d728d62aa5529c605afc0
-SIZE (bird-1.6.4.tar.gz) = 1018874
+TIMESTAMP = 1521724692
+SHA256 (bird-2.0.2.tar.gz) = 035f91d6f568f8ed438a0814235ac4c5c79147cd2acf201322c307732883480f
+SIZE (bird-2.0.2.tar.gz) = 1015224
Index: net/bird/files/bird.in
===================================================================
--- net/bird/files/bird.in	(revision 465813)
+++ net/bird/files/bird.in	(working copy)
@@ -2,7 +2,7 @@
 
 # $FreeBSD$
 #
-# PROVIDE: bird dynamicrouting
+# PROVIDE: bird
 # REQUIRE: LOGIN
 # KEYWORD: shutdown
 #
@@ -24,11 +24,9 @@
 
 load_rc_config $name
 
-extra_commands="reload"
-
 : ${bird_enable="NO"}
 : ${bird_config="%%PREFIX%%/etc/bird.conf"}
 
-command_args="-P /var/run/${name}.pid -c $bird_config"
+command_args="-c $bird_config"
 
 run_rc_command "$1"
Index: net/bird/files/bird6.in
===================================================================
--- net/bird/files/bird6.in	(revision 465813)
+++ net/bird/files/bird6.in	(nonexistent)
@@ -1,34 +0,0 @@
-#!/bin/sh
-
-# $FreeBSD$
-#
-# PROVIDE: bird6 dynamicrouting
-# REQUIRE: LOGIN
-# KEYWORD: shutdown
-#
-# Add the following lines to /etc/rc.conf.local or /etc/rc.conf
-# to enable this service:
-#
-# bird6_enable (bool):   Set to NO by default.
-#               Set it to YES to enable bird6.
-# bird6_config (path):   Set to %%PREFIX%%/etc/bird6.conf
-#               by default.
-#
-
-. /etc/rc.subr
-
-name="bird6"
-rcvar=bird6_enable
-
-command=%%PREFIX%%/sbin/${name}
-
-load_rc_config $name
-
-extra_commands="reload"
-
-: ${bird6_enable="NO"}
-: ${bird6_config="%%PREFIX%%/etc/bird6.conf"}
-
-command_args="-P /var/run/${name}.pid -c $bird6_config"
-
-run_rc_command "$1"

Property changes on: net/bird/files/bird6.in
___________________________________________________________________
Deleted: svn:eol-style
## -1 +0,0 ##
-native
\ No newline at end of property
Deleted: svn:keywords
## -1 +0,0 ##
-FreeBSD=%H
\ No newline at end of property
Deleted: svn:mime-type
## -1 +0,0 ##
-text/plain
\ No newline at end of property
Index: net/bird/files/patch-sysdep-bsd-setkey.h
===================================================================
--- net/bird/files/patch-sysdep-bsd-setkey.h	(revision 465813)
+++ net/bird/files/patch-sysdep-bsd-setkey.h	(nonexistent)
@@ -1,19 +0,0 @@
---- sysdep/bsd/setkey.h.orig	2018-03-22 12:32:46 UTC
-+++ sysdep/bsd/setkey.h
-@@ -158,12 +158,14 @@ sk_set_md5_in_sasp_db(sock *s, ip_addr local, ip_addr 
-     if (len > TCP_KEYLEN_MAX)
-       ERR_MSG("The password for TCP MD5 Signature is too long");
- 
--    if (setkey_md5(&src, &dst, passwd, SADB_ADD) < 0)
-+    if (setkey_md5(&src, &dst, passwd, SADB_ADD) < 0 ||
-+	setkey_md5(&dst, &src, passwd, SADB_ADD) < 0)
-       ERR_MSG("Cannot add TCP-MD5 password into the IPsec SA/SP database");
-   }
-   else
-   {
--    if (setkey_md5(&src, &dst, NULL, SADB_DELETE) < 0)
-+    if (setkey_md5(&src, &dst, NULL, SADB_DELETE) < 0 ||
-+	setkey_md5(&dst, &src, NULL, SADB_DELETE) < 0)
-       ERR_MSG("Cannot delete TCP-MD5 password from the IPsec SA/SP database");
-   }
-   return 0;

Property changes on: net/bird/files/patch-sysdep-bsd-setkey.h
___________________________________________________________________
Deleted: fbsd:nokeywords
## -1 +0,0 ##
-yes
\ No newline at end of property
Deleted: svn:eol-style
## -1 +0,0 ##
-native
\ No newline at end of property
Deleted: svn:mime-type
## -1 +0,0 ##
-text/plain
\ No newline at end of property
Index: net/bird/files/patch-sysdep_bsd_setkey.h
===================================================================
--- net/bird/files/patch-sysdep_bsd_setkey.h	(nonexistent)
+++ net/bird/files/patch-sysdep_bsd_setkey.h	(working copy)
@@ -0,0 +1,19 @@
+--- sysdep/bsd/setkey.h.orig	2018-01-16 09:46:07 UTC
++++ sysdep/bsd/setkey.h
+@@ -160,12 +160,14 @@ sk_set_md5_in_sasp_db(sock *s, ip_addr local, ip_addr 
+     if (len > TCP_KEYLEN_MAX)
+       ERR_MSG("The password for TCP MD5 Signature is too long");
+ 
+-    if (setkey_md5(&src, &dst, pxlen, passwd, SADB_ADD) < 0)
++    if (setkey_md5(&src, &dst, pxlen, passwd, SADB_ADD) < 0 ||
++		setkey_md5(&dst, &src, pxlen, passwd, SADB_ADD) < 0)
+       ERR_MSG("Cannot add TCP-MD5 password into the IPsec SA/SP database");
+   }
+   else
+   {
+-    if (setkey_md5(&src, &dst, pxlen, NULL, SADB_DELETE) < 0)
++    if (setkey_md5(&src, &dst, pxlen, NULL, SADB_DELETE) < 0 ||
++		setkey_md5(&dst, &src, pxlen, NULL, SADB_DELETE) < 0)
+       ERR_MSG("Cannot delete TCP-MD5 password from the IPsec SA/SP database");
+   }
+   return 0;

Property changes on: net/bird/files/patch-sysdep_bsd_setkey.h
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: net/bird/files/patch-tools-Makefile.in
===================================================================
--- net/bird/files/patch-tools-Makefile.in	(revision 465813)
+++ net/bird/files/patch-tools-Makefile.in	(nonexistent)
@@ -1,15 +0,0 @@
---- tools/Makefile.in.orig	2016-04-29 09:13:23 UTC
-+++ tools/Makefile.in
-@@ -77,11 +77,7 @@ install: all
- 	if test -n "@CLIENT@" ; then								\
- 		$(INSTALL_PROGRAM) $(exedir)/birdc $(DESTDIR)/$(sbindir)/birdc@SUFFIX@ ;	\
- 	fi
--	if ! test -f $(DESTDIR)/@CONFIG_FILE@ ; then						\
--		$(INSTALL_DATA) $(srcdir)/doc/bird.conf.example $(DESTDIR)/@CONFIG_FILE@ ;	\
--	else											\
--		echo "Not overwriting old bird@SUFFIX@.conf" ;					\
--	fi
-+	$(INSTALL_DATA) $(srcdir)/doc/bird.conf.example $(DESTDIR)/@CONFIG_FILE@.example
- 
- install-docs:
- 	$(INSTALL) -d $(DESTDIR)/$(docdir)

Property changes on: net/bird/files/patch-tools-Makefile.in
___________________________________________________________________
Deleted: fbsd:nokeywords
## -1 +0,0 ##
-yes
\ No newline at end of property
Deleted: svn:eol-style
## -1 +0,0 ##
-native
\ No newline at end of property
Deleted: svn:mime-type
## -1 +0,0 ##
-text/plain
\ No newline at end of property
Index: net/bird/files/patch-tools-Rules.in
===================================================================
--- net/bird/files/patch-tools-Rules.in	(revision 465813)
+++ net/bird/files/patch-tools-Rules.in	(nonexistent)
@@ -1,14 +0,0 @@
---- tools/Rules.in.orig	2016-04-29 09:13:23 UTC
-+++ tools/Rules.in
-@@ -66,10 +66,8 @@ else
- subdir: all.o
- 
- all.o: $(objs)
--#       $(LD) -r -o $@ $^
--#	Changed to $(CC) because $(LD) has problems with crosscompiling
- 	@echo LD -r -o $@ $^
--	@$(CC) -nostdlib -r -o $@ $^
-+	$(LD) -r -o $@ $^
- 
- endif
- 

Property changes on: net/bird/files/patch-tools-Rules.in
___________________________________________________________________
Deleted: fbsd:nokeywords
## -1 +0,0 ##
-yes
\ No newline at end of property
Deleted: svn:eol-style
## -1 +0,0 ##
-native
\ No newline at end of property
Deleted: svn:mime-type
## -1 +0,0 ##
-text/plain
\ No newline at end of property
Index: net/bird/files/pkg-message.in
===================================================================
--- net/bird/files/pkg-message.in	(nonexistent)
+++ net/bird/files/pkg-message.in	(working copy)
@@ -0,0 +1,9 @@
+#######################################################################
+
+%%LOCALBASE%%/sbin/birdc was added to /etc/shells
+To assign it to a user, the user needs to be in group wheel.
+
+Alternatively you could use security/sudo to grant a user access to it,
+which would allow '%%LOCALBASE%%/sbin/birdc -r' to restrict to read-only.
+
+#######################################################################

Property changes on: net/bird/files/pkg-message.in
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: net/bird/pkg-descr
===================================================================
--- net/bird/pkg-descr	(revision 465813)
+++ net/bird/pkg-descr	(working copy)
@@ -5,6 +5,8 @@
 - BGP
 - RIP
 - OSPF
+- LDP
+- L3VPN
 - Static routes
 - Inter-table protocol
 - Command-line interface
Index: net/bird/pkg-plist
===================================================================
--- net/bird/pkg-plist	(revision 465813)
+++ net/bird/pkg-plist	(working copy)
@@ -1,4 +1,4 @@
-@sample etc/bird%%VER%%.conf.example etc/bird%%VER%%.conf
-sbin/bird%%VER%%
-sbin/birdc%%VER%%
-sbin/birdcl%%VER%%
+@sample etc/bird.conf.sample
+@shell sbin/birdc
+sbin/bird
+sbin/birdcl
