Index: archivers/liblz4/Makefile
===================================================================
--- archivers/liblz4/Makefile	(révision 469434)
+++ archivers/liblz4/Makefile	(copie de travail)
@@ -17,6 +17,7 @@
 USES=		gmake pkgconfig
 USE_GITHUB=	yes
 USE_LDCONFIG=	yes
+MAKE_ENV+=	TARGET_ARCH=
 MAKE_ARGS=	PREFIX="${PREFIX}" LIBDIR="${PREFIX}/lib" \
 		INSTALL_PROGRAM="${INSTALL_PROGRAM}" \
 		MANDIR="${MAN1PREFIX}/man"
