Index: net/bird/Makefile
===================================================================
--- net/bird/Makefile	(revision 440546)
+++ net/bird/Makefile	(working copy)
@@ -3,7 +3,7 @@
 
 PORTNAME?=	bird
 PORTVERSION=	1.6.3
-PORTREVISION=	1
+PORTREVISION=	2
 CATEGORIES=	net
 MASTER_SITES=	ftp://bird.network.cz/pub/bird/ \
 		http://bird.mpls.in/distfiles/bird/
Index: net/bird/files/bird.in
===================================================================
--- net/bird/files/bird.in	(revision 440546)
+++ net/bird/files/bird.in	(working copy)
@@ -29,6 +29,6 @@
 : ${bird_enable="NO"}
 : ${bird_config="%%PREFIX%%/etc/bird.conf"}
 
-command_args="-c $bird_config"
+command_args="-P /var/run/${name}.pid -c $bird_config"
 
 run_rc_command "$1"
Index: net/bird/files/bird6.in
===================================================================
--- net/bird/files/bird6.in	(revision 440546)
+++ net/bird/files/bird6.in	(working copy)
@@ -29,6 +29,6 @@
 : ${bird6_enable="NO"}
 : ${bird6_config="%%PREFIX%%/etc/bird6.conf"}
 
-command_args="-c $bird6_config"
+command_args="-P /var/run/${name}.pid -c $bird6_config"
 
 run_rc_command "$1"
Index: net/bird/files/firewall_support.patch
===================================================================
--- net/bird/files/firewall_support.patch	(revision 440546)
+++ net/bird/files/firewall_support.patch	(working copy)
@@ -161,7 +161,7 @@
 +CF_ADDTO(proto, firewall_proto '}')
 +
 +firewall_proto_start: proto_start FIREWALL {
-+     this_proto = proto_config_new(&proto_firewall, sizeof(struct firewall_config), $1);
++     this_proto = proto_config_new(&proto_firewall, $1);
 +     this_proto->preference = 0;
 +     FIREWALL_CFG->flush_start = 1;
 +     FIREWALL_CFG->flush_shutdown = 1;
@@ -217,7 +217,7 @@
 index 0000000..e447470
 --- /dev/null
 +++ proto/firewall/firewall.c
-@@ -0,0 +1,198 @@
+@@ -0,0 +1,199 @@
 +/*
 + *	BIRD -- Firewall Protocol Configuration
 + *
@@ -247,7 +247,7 @@
 +#include "firewall.h"
 +
 +static int init_done = 0;
-+struct rate_limit rl_fw_err;
++struct tbf rl_fw_err;
 +
 +static void
 +firewall_collect(void)
@@ -408,6 +408,7 @@
 +  name:		"Firewall",
 +  template:	"fw%d",
 +  attr_class:	EAP_FIREWALL,
++  config_size:	sizeof(struct firewall_config),
 +  init:		firewall_init,
 +  start:	firewall_start,
 +  shutdown:	firewall_shutdown,
@@ -472,7 +473,7 @@
 +extern struct protocol proto_firewall;
 +
 +extern struct firewall_control fw_ipfw, fw_pf, fw_ipset;
-+extern struct rate_limit rl_fw_err;
++extern struct tbf rl_fw_err;
 +#define FW_ERR(x, y...)	log_rl(&rl_fw_err, L_ERR x, ##y)
 +
 +#endif
Index: net/bird/files/patch-setkey.h
===================================================================
--- net/bird/files/patch-setkey.h	(nonexistent)
+++ net/bird/files/patch-setkey.h	(working copy)
@@ -0,0 +1,17 @@
+--- sysdep/bsd/setkey.h.orig	2017-04-28 12:21:39.616325000 +0300
++++ sysdep/bsd/setkey.h	2017-04-28 12:20:57.712621000 +0300
+@@ -160,10 +160,14 @@ sk_set_md5_in_sasp_db(sock *s, ip_addr local, ip_addr 
+ 
+     if (setkey_md5(&src, &dst, passwd, SADB_ADD) < 0)
+       ERR_MSG("Cannot add TCP-MD5 password into the IPsec SA/SP database");
++    if (setkey_md5(&dst, &src, passwd, SADB_ADD) < 0)
++      ERR_MSG("Cannot add TCP-MD5 password into the IPsec SA/SP database");
+   }
+   else
+   {
+     if (setkey_md5(&src, &dst, NULL, SADB_DELETE) < 0)
++      ERR_MSG("Cannot delete TCP-MD5 password from the IPsec SA/SP database");
++    if (setkey_md5(&dst, &src, NULL, SADB_DELETE) < 0)
+       ERR_MSG("Cannot delete TCP-MD5 password from the IPsec SA/SP database");
+   }
+   return 0;

Property changes on: net/bird/files/patch-setkey.h
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
