diff --git net/pimd/Makefile net/pimd/Makefile
index f4e8c572d5af..d90267d63b51 100644
--- net/pimd/Makefile
+++ net/pimd/Makefile
@@ -2,6 +2,7 @@ PORTNAME=	pimd
 PORTVERSION=	2.3.2
 PORTREVISION=	1
 CATEGORIES=	net
+PORTREVISION=  1
 
 MAINTAINER=	olivier@FreeBSD.org
 COMMENT=	Lightweight stand-alone PIM-SM v2 multicast routing daemon
@@ -19,7 +20,7 @@ GH_SUBDIR=	libite:libite
 USE_RC_SUBR=	${PORTNAME}
 
 HAS_CONFIGURE=	yes
-MAKE_ARGS=	prefix="${PREFIX}" sysconfdir="${PREFIX}/etc/"
+MAKE_ARGS=	prefix="${PREFIX}" sysconfdir="${PREFIX}/etc"
 
 do-install:
 	${INSTALL_PROGRAM} ${WRKSRC}/pimd ${STAGEDIR}${PREFIX}/sbin/
diff --git net/pimd/files/pimd.in net/pimd/files/pimd.in
index d3339b5e9c70..505ecccb6ac8 100644
--- net/pimd/files/pimd.in
+++ net/pimd/files/pimd.in
@@ -2,7 +2,7 @@
 #
 # PROVIDE: pimd
 # REQUIRE: netif routing
-# KEYWORD: nojail
+# KEYWORD: nojailvnet
 
 #
 # Add the following to /etc/rc.conf[.local] to enable this service
@@ -17,6 +17,7 @@ rcvar="${name}_enable"
 command="%%PREFIX%%/sbin/${name}"
 required_files="%%PREFIX%%/etc/${name}.conf"
 extra_commands=reload
+required_modules="ip_mroute"
 
 load_rc_config $name
 : ${pimd_enable="NO"}
