Index: net/bird/Makefile
===================================================================
--- net/bird/Makefile	(revision 414946)
+++ net/bird/Makefile	(working copy)
@@ -2,8 +2,7 @@
 # $FreeBSD$
 
 PORTNAME?=	bird
-PORTVERSION=	1.5.0
-PORTREVISION=	1
+PORTVERSION=	1.6.0
 CATEGORIES=	net
 MASTER_SITES=	ftp://bird.network.cz/pub/bird/ \
 		http://bird.mpls.in/distfiles/bird/
Index: net/bird/distinfo
===================================================================
--- net/bird/distinfo	(revision 414946)
+++ net/bird/distinfo	(working copy)
@@ -1,2 +1,2 @@
-SHA256 (bird-1.5.0.tar.gz) = 79d5c4254e25f5890626830ca42aae03735716bdf152ddd8368473d2b3c17b5d
-SIZE (bird-1.5.0.tar.gz) = 1300141
+SHA256 (bird-1.6.0.tar.gz) = db3dfe37deac73b38b7d6966e577985f7b8e23f246760cb095ad13b0d8010e13
+SIZE (bird-1.6.0.tar.gz) = 1273899
Index: net/bird/files/bird.in
===================================================================
--- net/bird/files/bird.in	(revision 414946)
+++ net/bird/files/bird.in	(working copy)
@@ -29,6 +29,6 @@
 : ${bird_enable="NO"}
 : ${bird_config="%%PREFIX%%/etc/bird.conf"}
 
-command_args="-c $bird_config"
+command_args="-P /var/run/${name}.pid -c $bird_config"
 
 run_rc_command "$1"
Index: net/bird/files/bird6.in
===================================================================
--- net/bird/files/bird6.in	(revision 414946)
+++ net/bird/files/bird6.in	(working copy)
@@ -29,6 +29,6 @@
 : ${bird6_enable="NO"}
 : ${bird6_config="%%PREFIX%%/etc/bird6.conf"}
 
-command_args="-c $bird6_config"
+command_args="-P /var/run/${name}.pid -c $bird6_config"
 
 run_rc_command "$1"
Index: net/bird/files/patch-tools-Rules.in
===================================================================
--- net/bird/files/patch-tools-Rules.in	(revision 414946)
+++ net/bird/files/patch-tools-Rules.in	(nonexistent)
@@ -1,13 +0,0 @@
---- tools/Rules.in.orig	2010-10-10 21:48:20.000000000 +0200
-+++ tools/Rules.in	2010-12-18 15:00:27.000000000 +0100
-@@ -66,9 +66,7 @@
- subdir: all.o
- 
- all.o: $(objs)
--#       $(LD) -r -o $@ $^
--#	Changed to $(CC) because $(LD) has problems with crosscompiling
--	$(CC) -nostdlib -r -o $@ $^
-+	$(LD) -r -o $@ $^
- 
- endif
- 

Property changes on: net/bird/files/patch-tools-Rules.in
___________________________________________________________________
Deleted: fbsd:nokeywords
## -1 +0,0 ##
-yes
\ No newline at end of property
Deleted: svn:eol-style
## -1 +0,0 ##
-native
\ No newline at end of property
Deleted: svn:mime-type
## -1 +0,0 ##
-text/plain
\ No newline at end of property
