Index: sysutils/mstflint/Makefile
===================================================================
--- sysutils/mstflint/Makefile	(revision 522458)
+++ sysutils/mstflint/Makefile	(working copy)
@@ -2,7 +2,7 @@
 
 PORTNAME=	mstflint
 DISTVERSIONPREFIX=	v
-DISTVERSION=	4.12.0-1
+DISTVERSION=	4.13.3-2
 CATEGORIES=	sysutils
 
 MAINTAINER=	slavash@FreeBSD.org
@@ -11,11 +11,91 @@
 LICENSE=	GPLv2
 LICENSE_FILE=	${WRKSRC}/LICENSE
 
-USES=		autoreconf libtool
+USES=		autoreconf libtool gmake
 
+OPTIONS_DEFINE=	FWMANAGER ADAB STATIC
+OPTIONS_SUB=	yes
+
+.if defined(LITE)
+OPTIONS_SLAVE=	FWMANAGER ADAB STATIC
+.endif
+
+FWMANAGER_DESC=	Build with mstfwmanager and mstarchive
+ADAB_DESC=	Build with mstreg and mstlink
+STATIC_DESC=	Static binary (disable curl support)
+STATIC_CONFIGURE_ON=	--enable-all-static --disable-dc \
+			--disable-openssl --disable-inband
+
+LDFLAGS+=	-L${LOCALBASE}/lib
+CPPFLAGS+=	-I${LOCALBASE}/include
+
 GNU_CONFIGURE=	yes
 
 USE_GITHUB=	yes
 GH_ACCOUNT=	Mellanox
 
+.include <bsd.port.options.mk>
+
+.if ${PORT_OPTIONS:MSTATIC}
+.else
+USES+=	ssl
+.endif
+
+.if ${PORT_OPTIONS:MADAB}
+CONFIGURE_ARGS+=	--enable-adb-generic-tools
+.if ${PORT_OPTIONS:MSTATIC}
+BUILD_DEPENDS+=	xmlwf:textproc/expat2 \
+		boost-libs>0:devel/boost-libs
+.else
+LIB_DEPENDS+=	libexpat.so:textproc/expat2 \
+		libboost_regex.so:devel/boost-libs
+.endif
+.endif
+
+.if ${PORT_OPTIONS:MFWMANAGER}
+CONFIGURE_ARGS+=	--enable-fw-mgr
+.if ${PORT_OPTIONS:MSTATIC}
+BUILD_DEPENDS+=	xmllint:textproc/libxml2 \
+		${LOCALBASE}/include/boost/algorithm/string/detail/formatter_regex.hpp:devel/boost-libs
+EXTRA_PATCHES+=	${PATCHDIR}/extrat-patch-configure.ac \
+		${PATCHDIR}/extrat-patch-mlxarchive_Makefile.am \
+		${PATCHDIR}/extrat-patch-mlxarchive_mlxarchive_mfa2.cpp \
+		${PATCHDIR}/extrat-patch-mlxfwupdate_Makefile.am \
+		${PATCHDIR}/extrat-patch-mlxfwupdate_server_request.cpp
+.else
+LIB_DEPENDS+=	libxml2.so:textproc/libxml2 \
+		libboost_regex.so:devel/boost-libs \
+		libcurl.so:ftp/curl
+RUN_DEPENDS+=	ca_root_nss>=0:security/ca_root_nss
+.endif
+.endif
+
+.if defined(LITE)
+EXTRA_PATCHES+=	${PATCHDIR}/extrat-LITE-patch-Makefile.am \
+		${PATCHDIR}/extrat-LITE-patch-mlxconfig_Makefile.am \
+		${PATCHDIR}/extrat-LITE-patch-mlxreg_Makefile.am \
+		${PATCHDIR}/extrat-LITE-patch-cmdif_Makefile.am \
+		${PATCHDIR}/extrat-LITE-patch-mtcr_freebsd_Makefile.am \
+		${PATCHDIR}/extrat-LITE-patch-tools_layouts_Makefile.am \
+		${PATCHDIR}/extrat-LITE-patch-common_Makefile.am \
+		${PATCHDIR}/extrat-LITE-patch-dev_mgt_Makefile.am \
+		${PATCHDIR}/extrat-LITE-patch-reg_access_Makefile.am
+post-install:
+	${RM} ${STAGEDIR}${PREFIX}/lib/libadb_parser.a \
+	${STAGEDIR}${PREFIX}/lib/libcmdif.a \
+	${STAGEDIR}${PREFIX}/lib/libdev_mgt.a \
+	${STAGEDIR}${PREFIX}/lib/libmodules_lib.a \
+	${STAGEDIR}${PREFIX}/lib/libmstreg_lib.a \
+	${STAGEDIR}${PREFIX}/lib/libprint_util_lib.a \
+	${STAGEDIR}${PREFIX}/lib/libreg_access.a \
+	${STAGEDIR}${PREFIX}/lib/libtools_layouts.a \
+	${STAGEDIR}${PREFIX}/lib/mstflint/libmtcr_ul.a
+	${RMDIR} ${STAGEDIR}${PREFIX}/lib/mstflint
+.else
+post-install:
+	@${STRIP_CMD} ${STAGEDIR}${PREFIX}/lib/mstflint/python_tools/rreg_access.so
+	@${STRIP_CMD} ${STAGEDIR}${PREFIX}/lib/mstflint/python_tools/ccmdif.so
+	@${STRIP_CMD} ${STAGEDIR}${PREFIX}/lib/mstflint/python_tools/cmtcr.so
+	@${STRIP_CMD} ${STAGEDIR}${PREFIX}/lib/mstflint/python_tools/c_dev_mgt.so
+.endif
 .include <bsd.port.mk>
Index: sysutils/mstflint/distinfo
===================================================================
--- sysutils/mstflint/distinfo	(revision 522458)
+++ sysutils/mstflint/distinfo	(working copy)
@@ -1,3 +1,3 @@
-TIMESTAMP = 1562536530
-SHA256 (Mellanox-mstflint-v4.12.0-1_GH0.tar.gz) = 5cbf8924f33f813dd636bc25681b07d8940aa7803ff1f995266ac86ff30cbf35
-SIZE (Mellanox-mstflint-v4.12.0-1_GH0.tar.gz) = 3716706
+TIMESTAMP = 1578500867
+SHA256 (Mellanox-mstflint-v4.13.3-2_GH0.tar.gz) = 8bed4f44f857d7e5acce7100f6a402c2571e47778a8f66ab419a77bc12788ae6
+SIZE (Mellanox-mstflint-v4.13.3-2_GH0.tar.gz) = 3809937
Index: sysutils/mstflint/files/extrat-LITE-patch-Makefile.am
===================================================================
--- sysutils/mstflint/files/extrat-LITE-patch-Makefile.am	(nonexistent)
+++ sysutils/mstflint/files/extrat-LITE-patch-Makefile.am	(working copy)
@@ -0,0 +1,16 @@
+--- Makefile.am.orig	2020-01-05 16:50:06.000000000 +0100
++++ Makefile.am	2020-01-10 20:11:11.228348000 +0100
+@@ -30,11 +30,11 @@
+ # SOFTWARE.
+ #--
+ 
+-SUBDIRS = common mft_utils ext_libs $(TOOLS_CRYPTO) tools_layouts ${MTCR_CONF_DIR} mtcr_py $(MAD_IFC) reg_access cmdif $(XZ_UTILS_DIR) dev_mgt tools_res_mgmt mvpd mflash fw_comps_mgr libmfa pldmlib mlxconfig mlxfwops cmdparser $(FW_MGR_TOOLS) flint small_utils mstdump ${ADABE_TOOLS} tracers resourcedump
++SUBDIRS = mft_utils ext_libs $(TOOLS_CRYPTO) tools_layouts ${MTCR_CONF_DIR} $(MAD_IFC) reg_access cmdif $(XZ_UTILS_DIR) dev_mgt tools_res_mgmt mvpd mflash fw_comps_mgr libmfa pldmlib mlxconfig mlxfwops cmdparser mlxfwupdate ${ADABE_TOOLS}
+ 
+ DIST_SUBDIRS = tracers
+ 
+-man_MANS = man/mstflint.1 man/mstconfig.1 man/mstmcra.1 man/mstmread.1 man/mstmwrite.1 man/mstmtserver.1 man/mstregdump.1 man/mstvpd.1 man/mstprivhost.1 man/mstarchive.1 man/mstreg.1 man/mstfwtrace.1 man/mstlink.1 man/mstcongestion.1 man/mstfwmanager.1 man/mstfwreset.1 man/mstresourcedump.1
++man_MANS = man/mstconfig.1 man/mstlink.1 man/mstfwmanager.1
+ 
+ EXTRA_DIST = \
+ 	mstflint.spec \

Property changes on: sysutils/mstflint/files/extrat-LITE-patch-Makefile.am
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-LITE-patch-cmdif_Makefile.am
===================================================================
--- sysutils/mstflint/files/extrat-LITE-patch-cmdif_Makefile.am	(nonexistent)
+++ sysutils/mstflint/files/extrat-LITE-patch-cmdif_Makefile.am	(working copy)
@@ -0,0 +1,23 @@
+--- cmdif/Makefile.am.orig	2020-01-10 21:59:19.228661000 +0100
++++ cmdif/Makefile.am	2020-01-10 22:51:51.467764000 +0100
+@@ -35,7 +35,6 @@ USER_DIR = $(top_srcdir)
+ MTCR_DIR = $(USER_DIR)/include/mtcr_ul
+ TOOLS_LAYOUTS_DIR = $(USER_DIR)/tools_layouts
+ AM_CPPFLAGS = -I. -I../common -I../tools_layouts  -I${MTCR_DIR} -I.. -I$(USER_DIR)/${MTCR_CONF_DIR}
+-CCMDIF_SO = ccmdif.so
+ 
+ AM_CFLAGS = -W -Wall -g -MP -MD $(COMPILER_FPIC) -DCMDIF_EXPORTS
+ CMDIF_VERSION = 1
+@@ -43,12 +42,8 @@ lib_LTLIBRARIES = libcmdif.a
+ 
+ cmdifincludedir = $(includedir)/mstflint/cmdif/
+ 
+-cmdifinclude_HEADERS = icmd_cif_common.h icmd_cif_open.h
+-
+ libcmdif_a_SOURCES = tools_cif.c tools_cif.h icmd_cif_common.c icmd_cif_common.h icmd_cif_open.c icmd_cif_open.h
+ 
+-cmdif_pylibdir = $(libdir)/mstflint/python_tools/
+-cmdif_pylib_DATA = ${CCMDIF_SO} cmdif.py
+ ${CCMDIF_SO}: libcmdif.a
+ 	$(CC) -g -Wall -pthread -shared ${CFLAGS} *.o -o ${CCMDIF_SO} \
+ 	-L$(USER_DIR)/tools_layouts -ltools_layouts -L$(USER_DIR)/${MTCR_CONF_DIR} -lmtcr_ul

Property changes on: sysutils/mstflint/files/extrat-LITE-patch-cmdif_Makefile.am
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-LITE-patch-common_Makefile.am
===================================================================
--- sysutils/mstflint/files/extrat-LITE-patch-common_Makefile.am	(nonexistent)
+++ sysutils/mstflint/files/extrat-LITE-patch-common_Makefile.am	(working copy)
@@ -0,0 +1,11 @@
+--- common/Makefile.am.orig	2020-01-10 22:31:52.576485000 +0100
++++ common/Makefile.am	2020-01-10 22:32:24.068225000 +0100
+@@ -36,8 +36,6 @@ noinst_HEADERS=compatibility.h bit_slice.h tools_utils
+ 
+ commonincludedir = $(includedir)/mstflint/common/
+ 
+-commoninclude_HEADERS = compatibility.h
+-
+ update_prefix = sed -e 's,[@]MST_LIB_DIR[@]${CONF_DISABLE_PATH_UPDATE},$(libdir),g'\
+                     -e 's,[@]MST_BIN_DIR[@]${CONF_DISABLE_PATH_UPDATE},$(bindir),g'\
+                     -e 's,[@]MFTCONF_PREFIX[@],$(prefix),g'

Property changes on: sysutils/mstflint/files/extrat-LITE-patch-common_Makefile.am
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-LITE-patch-dev_mgt_Makefile.am
===================================================================
--- sysutils/mstflint/files/extrat-LITE-patch-dev_mgt_Makefile.am	(nonexistent)
+++ sysutils/mstflint/files/extrat-LITE-patch-dev_mgt_Makefile.am	(working copy)
@@ -0,0 +1,16 @@
+*** dev_mgt/Makefile.am.orig	Fri Jan 10 23:00:25 2020
+--- dev_mgt/Makefile.am	Fri Jan 10 23:18:29 2020
+*************** libdev_mgt_a_SOURCES = \
+*** 44,54 ****
+  						$(top_srcdir)/common/compatibility.h   \
+  						$(top_srcdir)/common/tools_version.h
+  
+- dev_mgt_pylibdir = $(libdir)/mstflint/python_tools/
+- 
+- dev_mgt_pylib_DATA = c_dev_mgt.so dev_mgt.py
+- dist_dev_mgt_pylib_DATA = dev_mgt.py
+- 
+  c_dev_mgt.so: libdev_mgt.a
+  	$(CC) -g -Wall -pthread -shared ${CFLAGS} tools_dev_types.o -o c_dev_mgt.so \
+  	-L$(USER_DIR)/reg_access -lreg_access \
+--- 44,49 ----

Property changes on: sysutils/mstflint/files/extrat-LITE-patch-dev_mgt_Makefile.am
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-LITE-patch-mlxconfig_Makefile.am
===================================================================
--- sysutils/mstflint/files/extrat-LITE-patch-mlxconfig_Makefile.am	(nonexistent)
+++ sysutils/mstflint/files/extrat-LITE-patch-mlxconfig_Makefile.am	(working copy)
@@ -0,0 +1,16 @@
+--- mlxconfig/Makefile.am.orig	2020-01-10 20:19:15.456616000 +0100
++++ mlxconfig/Makefile.am	2020-01-10 20:19:48.307817000 +0100
+@@ -55,13 +55,6 @@ AM_CPPFLAGS = -I. -I$(USER_DIR) -I$(top_srcdir)/includ
+ 
+ AM_CXXFLAGS = -pthread -Wall -W -g -MP -MD -pipe  -Wno-deprecated-declarations  $(COMPILER_FPIC) -DDATA_PATH=\"$(pkgdatadir)\"
+ bin_PROGRAMS = mstconfig
+-MLXPRIVHOST_PYTHON_WRAPPER=mstprivhost
+-bin_SCRIPTS = ${MLXPRIVHOST_PYTHON_WRAPPER}
+-${MLXPRIVHOST_PYTHON_WRAPPER}: $(PYTHON_WRAPPER_SCRIPT)
+-	cp $(PYTHON_WRAPPER_SCRIPT) $@
+-
+-mlxprivhostlibdir=$(libdir)/mstflint/python_tools/$(MLXPRIVHOST_PYTHON_WRAPPER)
+-mlxprivhostlib_SCRIPTS = $(MLXPRIVHOST_PYTHON_WRAPPER).py
+ 
+ noinst_LTLIBRARIES = libmlxcfg.a
+ 

Property changes on: sysutils/mstflint/files/extrat-LITE-patch-mlxconfig_Makefile.am
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-LITE-patch-mlxreg_Makefile.am
===================================================================
--- sysutils/mstflint/files/extrat-LITE-patch-mlxreg_Makefile.am	(nonexistent)
+++ sysutils/mstflint/files/extrat-LITE-patch-mlxreg_Makefile.am	(working copy)
@@ -0,0 +1,11 @@
+--- mlxreg/Makefile.am.orig	2020-01-10 20:34:08.906577000 +0100
++++ mlxreg/Makefile.am	2020-01-10 20:34:18.747595000 +0100
+@@ -58,8 +58,6 @@ libmstreg_lib_a_SOURCES = mlxreg_exception.h \
+ 
+ mstreg_SOURCES = mlxreg_ui.cpp mlxreg_ui.h
+ 
+-bin_PROGRAMS = mstreg
+-
+ mstreg_LDADD = libmstreg_lib.a \
+                 $(USER_DIR)/cmdparser/libcmdparser.a \
+                 $(MFT_UTILS_DIR)/libmftutils.a \

Property changes on: sysutils/mstflint/files/extrat-LITE-patch-mlxreg_Makefile.am
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-LITE-patch-mtcr_freebsd_Makefile.am
===================================================================
--- sysutils/mstflint/files/extrat-LITE-patch-mtcr_freebsd_Makefile.am	(nonexistent)
+++ sysutils/mstflint/files/extrat-LITE-patch-mtcr_freebsd_Makefile.am	(working copy)
@@ -0,0 +1,8 @@
+--- mtcr_freebsd/Makefile.am.orig	2020-01-10 22:04:45.957050000 +0100
++++ mtcr_freebsd/Makefile.am	2020-01-10 22:04:59.716184000 +0100
+@@ -50,5 +50,3 @@ libmtcr_ul_a_SOURCES += ../mtcr_ul/mtcr_ib_ofed.c
+ endif
+ 
+ libraryincludedir=$(includedir)/mstflint
+-libraryinclude_HEADERS = $(top_srcdir)/include/mtcr_ul/mtcr.h  $(top_srcdir)/include/mtcr_ul/mtcr_com_defs.h
+-

Property changes on: sysutils/mstflint/files/extrat-LITE-patch-mtcr_freebsd_Makefile.am
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-LITE-patch-reg_access_Makefile.am
===================================================================
--- sysutils/mstflint/files/extrat-LITE-patch-reg_access_Makefile.am	(nonexistent)
+++ sysutils/mstflint/files/extrat-LITE-patch-reg_access_Makefile.am	(working copy)
@@ -0,0 +1,12 @@
+--- reg_access/Makefile.am.orig	2020-01-05 16:50:06.000000000 +0100
++++ reg_access/Makefile.am	2020-01-10 23:33:56.588295000 +0100
+@@ -49,9 +49,6 @@ libreg_access_a_LIBADD = $(libreg_access_a_DEPENDENCIE
+ RREG_ACCESS_SO = rreg_access.so
+ reg_access_pylibdir = $(libdir)/mstflint/python_tools/
+ 
+-reg_access_pylib_DATA = ${RREG_ACCESS_SO} regaccess.py
+-dist_reg_access_pylib_DATA = regaccess.py
+-
+ ${RREG_ACCESS_SO}: libreg_access.a
+ 	$(CC) -g -Wall -pthread -shared ${CFLAGS} *.o -o ${RREG_ACCESS_SO} \
+ 	-L$(USER_DIR)/tools_layouts -ltools_layouts -L$(MTCR_DIR) -lmtcr_ul

Property changes on: sysutils/mstflint/files/extrat-LITE-patch-reg_access_Makefile.am
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-LITE-patch-tools_layouts_Makefile.am
===================================================================
--- sysutils/mstflint/files/extrat-LITE-patch-tools_layouts_Makefile.am	(nonexistent)
+++ sysutils/mstflint/files/extrat-LITE-patch-tools_layouts_Makefile.am	(working copy)
@@ -0,0 +1,11 @@
+--- tools_layouts/Makefile.am.orig	2020-01-10 22:06:24.040554000 +0100
++++ tools_layouts/Makefile.am	2020-01-10 22:06:44.128071000 +0100
+@@ -41,8 +41,6 @@ lib_LTLIBRARIES = libtools_layouts.a
+ 
+ toolslayoutsincludedir = $(includedir)/mstflint/tools_layouts/
+ 
+-toolslayoutsinclude_HEADERS = icmd_layouts.h adb_to_c_utils.h
+-
+ libtools_layouts_a_SOURCES = prm_adb_db.cpp prm_adb_db.h \
+ 				adb_to_c_utils.c adb_to_c_utils.h\
+ 				cibfw_layouts.c cibfw_layouts.h \

Property changes on: sysutils/mstflint/files/extrat-LITE-patch-tools_layouts_Makefile.am
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-patch-configure.ac
===================================================================
--- sysutils/mstflint/files/extrat-patch-configure.ac	(nonexistent)
+++ sysutils/mstflint/files/extrat-patch-configure.ac	(working copy)
@@ -0,0 +1,36 @@
+--- configure.ac.orig	2020-01-05 16:50:06.000000000 +0100
++++ configure.ac	2020-01-09 00:15:46.179518000 +0100
+@@ -104,12 +104,10 @@ AC_ARG_ENABLE(fw-mgr,
+              [enable_fw_mgr="no"])
+ AC_MSG_RESULT($enable_fw_mgr)
+ if test "x$enable_fw_mgr" = "xyes"; then
+-    AC_CHECK_HEADER(curl/curl.h,,AC_MSG_ERROR([cannot find curl/curl.h . this header is needed for compiling fw manager tool]))
+     AC_CHECK_HEADER(zlib.h,,AC_MSG_ERROR([cannot find zlib.h . this header is needed for compiling fw manager tool]))
+     AC_CHECK_LIB(z, uncompress,, AC_MSG_ERROR([cannot find zlib uncompress() function.]))
+     AC_CHECK_HEADER(lzma.h,,AC_MSG_ERROR([Cannot find lzma.h.]))
+     AC_CHECK_LIB(boost_regex, regcompA,, AC_MSG_ERROR([cannot find boost_regex regcompA() function. Try re-installing the library...]))
+-    CURL_INC_DIR=/usr/include/curl/
+ fi
+ 
+ # Adabe
+@@ -133,8 +131,6 @@ if test "x$enable_adb_generic_tools" = "xyes"; then
+     AC_LANG_POP([C++])
+ fi
+ 
+-AC_SUBST(CURL_INC_DIR)
+-
+ AM_CONDITIONAL(ENABLE_FWMGR, [test "x$enable_fw_mgr" = "xyes" ])
+ AC_SUBST(ENABLE_FWMGR)
+ 
+@@ -352,11 +348,6 @@ AS_IF([test "x$enable_cs" = "xyes" || test "x$enable_o
+ AS_IF([test "x$enable_openssl" = "xyes"], [
+         AC_CONFIG_FILES(mlxsign_lib/Makefile)
+     ])
+-
+-if [ test "x$enable_openssl" = "xno" ] && \
+-    [ test "x$enable_fw_mgr" = "xyes" ] ; then
+-    AC_MSG_ERROR([--enable-fw-mgr cannot be used with --disabled-openssl.])
+-fi
+ 
+ AS_IF([test "x$enable_fw_mgr" = "xyes"], [
+         AC_CONFIG_FILES(mlxfwupdate/Makefile)

Property changes on: sysutils/mstflint/files/extrat-patch-configure.ac
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-patch-mlxarchive_Makefile.am
===================================================================
--- sysutils/mstflint/files/extrat-patch-mlxarchive_Makefile.am	(nonexistent)
+++ sysutils/mstflint/files/extrat-patch-mlxarchive_Makefile.am	(working copy)
@@ -0,0 +1,26 @@
+--- mlxarchive/Makefile.am.orig	2020-01-05 16:50:06.000000000 +0100
++++ mlxarchive/Makefile.am	2020-01-09 14:15:21.405697000 +0100
+@@ -20,8 +20,6 @@ DEV_MGT_DIR = $(USER_DIR)/dev_mgt
+ TOOLS_CRYPTO_DIR = $(USER_DIR)/tools_crypto
+ CMDIF_DIR = $(USER_DIR)/cmdif
+ 
+-CURL_LIBS = -lcurl  -lssl -lcrypto -lrt
+-
+ INCLUDES = $(JSON_CFLAGS) -I. \
+            -I$(USER_DIR) -I$(USER_DIR)/libmfa \
+            -I$(MTCR_DIR) -I$(USER_DIR)/include/mtcr_ul \
+@@ -64,13 +62,13 @@ mstarchive_LDADD = libmstarchive.a\
+                    $(USER_DIR)/xz_utils/libxz_utils.a\
+                    $(USER_DIR)/ext_libs/minixz/libminixz.a\
+ 		   $(USER_DIR)/xz_utils/libxz_utils.a \
+-                   $(CURL_LIBS)\
+                    -lboost_regex -lboost_filesystem -lboost_system \
+                    -llzma -lm $(LIBSTD_CPP) ${LDL}
+ 
+ if ENABLE_OPENSSL
+ mstarchive_LDADD += $(top_srcdir)/mlxsign_lib/libmlxsign.a -lcrypto
+ else
++AM_CXXFLAGS += -DNO_OPEN_SSL
+ endif
+ 
+ if ENABLE_CS

Property changes on: sysutils/mstflint/files/extrat-patch-mlxarchive_Makefile.am
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-patch-mlxarchive_mlxarchive_mfa2.cpp
===================================================================
--- sysutils/mstflint/files/extrat-patch-mlxarchive_mlxarchive_mfa2.cpp	(nonexistent)
+++ sysutils/mstflint/files/extrat-patch-mlxarchive_mlxarchive_mfa2.cpp	(working copy)
@@ -0,0 +1,29 @@
+--- mlxarchive/mlxarchive_mfa2.cpp.orig	2020-01-05 16:50:06.000000000 +0100
++++ mlxarchive/mlxarchive_mfa2.cpp	2020-01-09 14:20:54.186228000 +0100
+@@ -39,7 +39,9 @@
+  */
+ 
+ #include <xz_utils/xz_utils.h>
++#ifndef NO_OPEN_SSL
+ #include <mlxsign_lib/mlxsign_lib.h>
++#endif
+ #include "mlxarchive_mfa2.h"
+ #include "mlxarchive_mfa2_utils.h"
+ 
+@@ -120,6 +122,7 @@ void MFA2::pack(vector<u_int8_t>& buff)
+     xz_compress_crc32(9, componentsBlockBuff.data(), componentsBlockBuff.size(),
+             zippedComponentBlockBuff.data(), zippedSize);
+ 
++#ifndef NO_OPEN_SSL
+     //compute descriptors SHA256
+     vector<u_int8_t> descriptorsBuff;
+     packDescriptors(descriptorsBuff);
+@@ -136,7 +139,7 @@ void MFA2::pack(vector<u_int8_t>& buff)
+     mlxSignSHA256 << zippedComponentBlockBuff;
+     mlxSignSHA256.getDigest(digest);
+     _packageDescriptor.setSHA256(digest);
+-
++#endif
+ }
+ 
+ /*void MFA2::update(vector<u_int8_t>& buff)

Property changes on: sysutils/mstflint/files/extrat-patch-mlxarchive_mlxarchive_mfa2.cpp
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-patch-mlxfwupdate_Makefile.am
===================================================================
--- sysutils/mstflint/files/extrat-patch-mlxfwupdate_Makefile.am	(nonexistent)
+++ sysutils/mstflint/files/extrat-patch-mlxfwupdate_Makefile.am	(working copy)
@@ -0,0 +1,38 @@
+--- mlxfwupdate/Makefile.am.orig	2020-01-09 18:28:23.591163000 +0100
++++ mlxfwupdate/Makefile.am	2020-01-09 18:35:23.551040000 +0100
+@@ -45,10 +45,6 @@ MTCR_UL_LIB = $(MTCR_UL_DIR)/libmtcr_ul.a
+ XML_FLAGS = -DUSE_XML -DLIBXML_STATIC
+ XML_LIBS = -lxml2 $(ZLIB_LIB)
+ 
+-CURL_LIBS = -lcurl  -lssl -lcrypto -lrt
+-
+-CURL_FLAGS = -DCURL_STATICLIB
+-
+ common_INCLUDES =  -I$(USER_DIR)/common\
+             -I$(USER_DIR)/\
+             -I$(USER_DIR)/include/mtcr_ul\
+@@ -59,7 +55,6 @@ common_INCLUDES =  -I$(USER_DIR)/common\
+             $(JSON_CFLAGS)\
+             -I$(USER_DIR)/mlxfwops/lib\
+             -I$(USER_DIR)/pldmlib\
+-            -I$(CURL_INC_DIR)\
+             -I$(USER_DIR)/mft_utils\
+             -I$(USER_DIR)/dev_mgt
+ 
+@@ -100,7 +95,6 @@ common_LDADD  = $(USER_DIR)/dev_mgt/libdev_mgt.a\
+                         $(USER_DIR)/tools_layouts/libtools_layouts.a\
+                         $(USER_DIR)/fw_comps_mgr/libfw_comps_mgr.a\
+                         $(XML_LIBS)\
+-                        $(CURL_LIBS)\
+                         -lm $(LIBSTD_CPP) ${LDL}
+                         
+ if ENABLE_OPENSSL
+@@ -118,6 +112,6 @@ common_LDADD += $(top_srcdir)/mad_ifc/libmad_ifc.a
+ else
+ endif
+ 
+-mstfwmanager_CXXFLAGS =  -DMSTFLINT -DUSE_CURL $(AM_CXXFLAGS) $(CURL_FLAGS) $(XML_FLAGS) -I$(MTCR_UL_DIR) $(common_INCLUDES) 
++mstfwmanager_CXXFLAGS =  -DMSTFLINT $(AM_CXXFLAGS) $(XML_FLAGS) -I$(MTCR_UL_DIR) $(common_INCLUDES) 
+ mstfwmanager_SOURCES  = $(common_SOURCEES)
+-mstfwmanager_LDADD    = $(common_LDADD) $(MTCR_UL_LIB) -lz
++mstfwmanager_LDADD    = $(common_LDADD) $(MTCR_UL_LIB) -lz -llzma

Property changes on: sysutils/mstflint/files/extrat-patch-mlxfwupdate_Makefile.am
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/extrat-patch-mlxfwupdate_server_request.cpp
===================================================================
--- sysutils/mstflint/files/extrat-patch-mlxfwupdate_server_request.cpp	(nonexistent)
+++ sysutils/mstflint/files/extrat-patch-mlxfwupdate_server_request.cpp	(working copy)
@@ -0,0 +1,23 @@
+--- mlxfwupdate/server_request.cpp.orig	2020-01-09 00:45:10.200149000 +0100
++++ mlxfwupdate/server_request.cpp	2020-01-09 00:46:56.459627000 +0100
+@@ -43,9 +43,7 @@
+ 
+ 
+ using namespace std;
+-#ifndef USE_CURL
+-#define USE_CURL
+-#endif
++#undef USE_CURL
+ 
+ extern int abort_request;
+ 
+@@ -62,7 +60,9 @@ ServerRequest::ServerRequest(const char *url, const ch
+     _show_progress = show_progress;
+     _ceritifcate   = certificate;
+     _numberOfRetrials = numberOfRetrials;
++#ifdef USE_CURL
+     _headers       = NULL;
++#endif
+     if (proxy != NULL) {
+         if (proxy[0] != 0) {
+             _UseProxy = 1;

Property changes on: sysutils/mstflint/files/extrat-patch-mlxfwupdate_server_request.cpp
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/patch-configure.ac
===================================================================
--- sysutils/mstflint/files/patch-configure.ac	(nonexistent)
+++ sysutils/mstflint/files/patch-configure.ac	(working copy)
@@ -0,0 +1,65 @@
+--- configure.ac.orig	2020-01-05 16:50:06.000000000 +0100
++++ configure.ac	2020-01-09 00:28:38.419054000 +0100
+@@ -100,7 +100,7 @@ fi
+ AC_MSG_CHECKING(--enable-fw-mgr argument)
+ AC_ARG_ENABLE(fw-mgr,
+              [  --enable-fw-mgr        Enable compiling mstfwmanager tool and features],
+-             [enable_fw_mgr="yes"],
++             [enable_fw_mgr=$enableval],
+              [enable_fw_mgr="no"])
+ AC_MSG_RESULT($enable_fw_mgr)
+ if test "x$enable_fw_mgr" = "xyes"; then
+@@ -116,7 +116,7 @@ fi
+ AC_MSG_CHECKING(--enable-adb-generic-tools argument)
+ AC_ARG_ENABLE(adb-generic-tools,
+              [  --enable-adb-generic-tools        Enable compiling the following tools which depends on ADABE: mstreg and mstlink],
+-             [enable_adb_generic_tools="yes"],
++             [enable_adb_generic_tools=$enableval],
+              [enable_adb_generic_tools="no"])
+ AC_MSG_RESULT($enable_adb_generic_tools)
+ if test "x$enable_adb_generic_tools" = "xyes"; then
+@@ -143,7 +143,7 @@ AC_SUBST(ENABLE_FWMGR)
+ AC_MSG_CHECKING(--enable-xml2 argument)
+ AC_ARG_ENABLE(xml2,
+             [  --enable-xml2           Enable mstflint libxml2 dependent features],
+-            [enable_xml2="yes"],
++            [enable_xml2=$enableval],
+             [enable_xml2="no"])
+ AC_MSG_RESULT($enable_xml2)
+ if test [ "x$enable_xml2" = "xyes" ] || [ test "x$enable_fw_mgr" = "xyes" ]; then
+@@ -185,7 +185,7 @@ AM_CONDITIONAL(ENABLE_INBAND, [test  "x$enable_inband"
+ AC_MSG_CHECKING(--enable-cs argument)
+ AC_ARG_ENABLE(cs,
+             [  --enable-cs             Enable mstflint "checksum" command, dependant of openssl library],
+-            [enable_cs="yes"],
++            [enable_cs=$enableval],
+             [enable_cs="no"])
+ AC_MSG_RESULT($enable_cs)
+ if test "x$enable_cs" = "xyes"; then
+@@ -211,7 +211,7 @@ fi
+ AC_MSG_CHECKING(--enable-all-static argument)
+ AC_ARG_ENABLE([all_static],
+     AS_HELP_STRING([--enable-all-static], [Enable creating none dynamic executables]),
+-    [enable_all_static="yes"],
++    [enable_all_static=$enableval],
+     [enable_all_static="no"])
+ AC_MSG_RESULT($enable_all_static)
+ 
+@@ -237,7 +237,7 @@ AC_ARG_ENABLE([static_libstdcpp],
+     AS_HELP_STRING([--enable-static-libstdcpp], [Enable link static to libstdc++]))
+ 
+ AS_IF([test "x$enable_static_libstdcpp" = "xyes" -o "x$enable_all_static" = "xyes" ], [
+-  CXXFLAGS="$CXXFLAGS -static-libstdc++ -static-libgcc"
++  CXXFLAGS="$CXXFLAGS -static"
+ ])
+ 
+ AC_ARG_ENABLE([dynamic_ld],
+@@ -267,7 +267,7 @@ AS_IF([test "x$INIPARSER_SYSTEM_AVAILABLE" = "xyes" ],
+     CXXFLAGS="$CXXFLAGS -DHAVE_INI_PARSER"
+ ])
+ 
+-AC_SEARCH_LIBS([json_object_get], [json-c jason], [JSON_SYSTEM_AVAILABLE="yes"],[
++AC_SEARCH_LIBS([json_object_get], [jsoncpp], [JSON_SYSTEM_AVAILABLE="yes"],[
+     JSON_SYSTEM_AVAILABLE="no"
+     JSON_CFLAGS='-I$(top_srcdir)/ext_libs/json'
+     AC_SUBST(JSON_CFLAGS)

Property changes on: sysutils/mstflint/files/patch-configure.ac
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/patch-mlxfwupdate_Makefile.am
===================================================================
--- sysutils/mstflint/files/patch-mlxfwupdate_Makefile.am	(nonexistent)
+++ sysutils/mstflint/files/patch-mlxfwupdate_Makefile.am	(working copy)
@@ -0,0 +1,11 @@
+--- mlxfwupdate/Makefile.am.orig	2019-11-15 19:39:03 UTC
++++ mlxfwupdate/Makefile.am
+@@ -37,7 +37,7 @@ MTCR_DIR = $(USER_DIR)/${MTCR_CONFIG_DIR}
+ bin_PROGRAMS = mstfwmanager
+ 
+ docdir=$(INSTALL_BASEDIR)/etc/mstflint
+-dist_doc_DATA=certificate/ca-bundle.crt
++#dist_doc_DATA=certificate/ca-bundle.crt
+ 
+ MTCR_UL_DIR = $(USER_DIR)/${MTCR_CONF_DIR}
+ MTCR_UL_LIB = $(MTCR_UL_DIR)/libmtcr_ul.a

Property changes on: sysutils/mstflint/files/patch-mlxfwupdate_Makefile.am
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/files/patch-mlxfwupdate_cmd__line__params.cpp
===================================================================
--- sysutils/mstflint/files/patch-mlxfwupdate_cmd__line__params.cpp	(nonexistent)
+++ sysutils/mstflint/files/patch-mlxfwupdate_cmd__line__params.cpp	(working copy)
@@ -0,0 +1,14 @@
+--- mlxfwupdate/cmd_line_params.cpp.orig	2019-11-16 07:49:07 UTC
++++ mlxfwupdate/cmd_line_params.cpp
+@@ -97,9 +97,9 @@ CmdLineParams::CmdLineParams()
+ 
+ #else
+     #ifdef MSTFLINT
+-    certificate = (string)ROOT_PATH + "etc/mstflint/ca-bundle.crt";
++    certificate = (string)ROOT_PATH + "/etc/ssl/cert.pem";
+     #else
+-    certificate = (string)ROOT_PATH + "etc/mft/ca-bundle.crt";
++    certificate = (string)ROOT_PATH + "/etc/ssl/cert.pem";
+     #endif
+ #endif
+ }

Property changes on: sysutils/mstflint/files/patch-mlxfwupdate_cmd__line__params.cpp
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: sysutils/mstflint/pkg-plist
===================================================================
--- sysutils/mstflint/pkg-plist	(revision 522458)
+++ sysutils/mstflint/pkg-plist	(working copy)
@@ -1,14 +1,19 @@
+%%FWMANAGER%%bin/mstarchive
 bin/mstconfig
 bin/mstcongestion
 bin/mstflint
+%%FWMANAGER%%bin/mstfwmanager
 bin/mstfwreset
 bin/mstfwtrace
+%%ADAB%%bin/mstlink
 bin/mstmcra
 bin/mstmread
 bin/mstmtserver
 bin/mstmwrite
 bin/mstprivhost
+%%ADAB%%bin/mstreg
 bin/mstregdump
+bin/mstresourcedump
 bin/mstvpd
 include/mstflint/cmdif/icmd_cif_common.h
 include/mstflint/cmdif/icmd_cif_open.h
@@ -17,8 +22,13 @@
 include/mstflint/mtcr_com_defs.h
 include/mstflint/tools_layouts/adb_to_c_utils.h
 include/mstflint/tools_layouts/icmd_layouts.h
+%%ADAB%%lib/libadb_parser.a
 lib/libcmdif.a
 lib/libdev_mgt.a
+%%ADAB%%lib/libmodules_lib.a
+%%FWMANAGER%%lib/libmstarchive.a
+%%ADAB%%lib/libmstreg_lib.a
+%%ADAB%%lib/libprint_util_lib.a
 lib/libreg_access.a
 lib/libtools_layouts.a
 lib/mstflint/libmtcr_ul.a
@@ -31,6 +41,8 @@
 lib/mstflint/python_tools/mlxpci/binary_file.py
 lib/mstflint/python_tools/mlxpci/mlxpci_lib.py
 lib/mstflint/python_tools/mstfwreset/mlxfwresetlib/__init__.py
+lib/mstflint/python_tools/mstfwreset/mlxfwresetlib/cmd_reg_mfrl.py
+lib/mstflint/python_tools/mstfwreset/mlxfwresetlib/cmd_reg_mpcir.py
 lib/mstflint/python_tools/mstfwreset/mlxfwresetlib/logger.py
 lib/mstflint/python_tools/mstfwreset/mlxfwresetlib/mcra.py
 lib/mstflint/python_tools/mstfwreset/mlxfwresetlib/mlnx_peripheral_components.py
@@ -43,13 +55,50 @@
 lib/mstflint/python_tools/mstfwtrace/mstfwtrace.py
 lib/mstflint/python_tools/mstfwtrace/secure_fw_trace.py
 lib/mstflint/python_tools/mstprivhost/mstprivhost.py
+lib/mstflint/python_tools/mstresourcedump/__init__.py
+lib/mstflint/python_tools/mstresourcedump/commands/CommandFactory.py
+lib/mstflint/python_tools/mstresourcedump/commands/DumpCommand.py
+lib/mstflint/python_tools/mstresourcedump/commands/QueryCommand.py
+lib/mstflint/python_tools/mstresourcedump/commands/ResDumpCommand.py
+lib/mstflint/python_tools/mstresourcedump/commands/__init__.py
+lib/mstflint/python_tools/mstresourcedump/fetchers/CapabilityFetcher.py
+lib/mstflint/python_tools/mstresourcedump/fetchers/ResourceDumpFetcher.py
+lib/mstflint/python_tools/mstresourcedump/fetchers/__init__.py
+lib/mstflint/python_tools/mstresourcedump/filters/SegmentsFilter.py
+lib/mstflint/python_tools/mstresourcedump/filters/__init__.py
+lib/mstflint/python_tools/mstresourcedump/mstresourcedump.py
+lib/mstflint/python_tools/mstresourcedump/resource_data/DataPrinter.py
+lib/mstflint/python_tools/mstresourcedump/resource_data/DumpData.py
+lib/mstflint/python_tools/mstresourcedump/resource_data/QueryData.py
+lib/mstflint/python_tools/mstresourcedump/resource_data/__init__.py
+lib/mstflint/python_tools/mstresourcedump/segments/CommandSegment.py
+lib/mstflint/python_tools/mstresourcedump/segments/ErrorSegment.py
+lib/mstflint/python_tools/mstresourcedump/segments/InfoSegment.py
+lib/mstflint/python_tools/mstresourcedump/segments/MenuRecord.py
+lib/mstflint/python_tools/mstresourcedump/segments/MenuSegment.py
+lib/mstflint/python_tools/mstresourcedump/segments/NoticeSegment.py
+lib/mstflint/python_tools/mstresourcedump/segments/RefSegment.py
+lib/mstflint/python_tools/mstresourcedump/segments/ResourceSegment.py
+lib/mstflint/python_tools/mstresourcedump/segments/Segment.py
+lib/mstflint/python_tools/mstresourcedump/segments/SegmentCreator.py
+lib/mstflint/python_tools/mstresourcedump/segments/SegmentFactory.py
+lib/mstflint/python_tools/mstresourcedump/segments/TerminateSegment.py
+lib/mstflint/python_tools/mstresourcedump/segments/__init__.py
+lib/mstflint/python_tools/mstresourcedump/utils/Exceptions.py
+lib/mstflint/python_tools/mstresourcedump/utils/constants.py
+lib/mstflint/python_tools/mstresourcedump/validation/ArgToMenuVerifier.py
+lib/mstflint/python_tools/mstresourcedump/validation/CapabilityValidator.py
+lib/mstflint/python_tools/mstresourcedump/validation/__init__.py
 lib/mstflint/python_tools/mtcr.py
 lib/mstflint/python_tools/regaccess.py
 lib/mstflint/python_tools/rreg_access.so
 lib/mstflint/python_tools/tools_version.py
 man/man1/mstarchive.1.gz
+man/man1/mstcongestion.1.gz
 man/man1/mstconfig.1.gz
 man/man1/mstflint.1.gz
+man/man1/mstfwmanager.1.gz
+man/man1/mstfwreset.1.gz
 man/man1/mstfwtrace.1.gz
 man/man1/mstlink.1.gz
 man/man1/mstmcra.1.gz
@@ -59,6 +108,7 @@
 man/man1/mstprivhost.1.gz
 man/man1/mstreg.1.gz
 man/man1/mstregdump.1.gz
+man/man1/mstresourcedump.1.gz
 man/man1/mstvpd.1.gz
 %%DATADIR%%/BlueField.csv
 %%DATADIR%%/ConnectIB.csv
@@ -69,6 +119,7 @@
 %%DATADIR%%/ConnectX4LX.csv
 %%DATADIR%%/ConnectX5.csv
 %%DATADIR%%/ConnectX6.csv
+%%DATADIR%%/ConnectX6DX.csv
 %%DATADIR%%/InfiniScaleIV.csv
 %%DATADIR%%/Quantum.csv
 %%DATADIR%%/Spectrum.csv
@@ -77,3 +128,5 @@
 %%DATADIR%%/SwitchX.csv
 %%DATADIR%%/mlxconfig_dbs/mlxconfig_host.db
 %%DATADIR%%/mlxconfig_dbs/mlxconfig_switch.db
+%%ADAB%%%%DATADIR%%/prm_dbs/hca/ext/register_access_table.adb
+%%ADAB%%%%DATADIR%%/prm_dbs/switch/ext/register_access_table.adb
