Index: Mk/bsd.default-versions.mk
===================================================================
--- Mk/bsd.default-versions.mk	(revision 465867)
+++ Mk/bsd.default-versions.mk	(working copy)
@@ -77,7 +77,7 @@
 # Possible values: 5.6, 7.0, 7.1, 7.2
 PHP_DEFAULT?=		5.6
 # Possible values: 2.7, 3.4, 3.5, 3.6
-PYTHON_DEFAULT?=	2.7
+PYTHON_DEFAULT?=	3.6
 # Possible values: 2.7
 PYTHON2_DEFAULT?=	2.7
 # Possible values: 3.4, 3.5, 3.6
