Index: net/frr/Makefile
===================================================================
--- net/frr/Makefile	(revision 464399)
+++ net/frr/Makefile	(working copy)
@@ -1,7 +1,7 @@
 # $FreeBSD$
 
 PORTNAME=	frr
-PORTVERSION=	3.0.3
+PORTVERSION=	4.0
 DISTVERSIONPREFIX=	frr-
 CATEGORIES=	net ipv6
 
@@ -11,7 +11,8 @@
 LICENSE=	GPLv2
 LICENSE_FILE=	${WRKSRC}/COPYING
 
-BUILD_DEPENDS=	gawk:lang/gawk
+BUILD_DEPENDS=	gawk:lang/gawk \
+				python:lang/python
 LIB_DEPENDS=	libjson-c.so:devel/json-c
 
 RESTRICTED=	possible GPL violation of Quagga project
@@ -18,18 +19,18 @@
 
 CONFLICTS=	openbgpd openospfd zebra quagga-esr quagga
 
-USE_GITHUB=	yes
-GH_ACCOUNT=	FRRouting
-
 GNU_CONFIGURE=	yes
 INSTALL_TARGET=	install-strip
 USES=		autoreconf bison gmake pkgconfig libtool makeinfo readline
 USE_LDCONFIG=	yes
 
+USE_GITHUB=	yes
+GH_ACCOUNT=	FRRouting
+
 INFO=		frr
 
-OPTIONS_DEFINE=		FPM LDPD PAM SHELLACCESS TCPSOCKETS VTYSH MULTIPATH TCMALLOC
-OPTIONS_DEFAULT=	VTYSH
+OPTIONS_DEFINE=		FPM LDPD PAM SHELLACCESS TCPSOCKETS VTYSH MULTIPATH TCMALLOC RPKI
+OPTIONS_DEFAULT=	VTYSH RPKI
 OPTIONS_SUB=	yes
 # SNMP option build failed
 # checksum.c:79:21: warning: implicit declaration of function 'typeof' is
@@ -41,6 +42,7 @@
 LDPD_DESC=		Build ldpd
 MULTIPATH_DESC=		Enable multipath function
 PAM_DESC=		Use libpam for PAM support in vtysh
+RPKI_DESC=		Add BGP RPKI (RFC 6810) support
 SHELLACCESS_DESC=	Allow users to access shell/telnet/ssh
 SNMP_DESC=		SNMP support
 TCMALLOC_DESC=		Use tcmalloc
@@ -60,7 +62,8 @@
 		--enable-vty-group=frrvty \
 		--sysconfdir=${ETCDIR} \
 		--localstatedir=/var/run/frr \
-		--disable-pimd
+		--disable-pimd \
+		--disable-nhrpd
 
 SHELLACCESS_CONFIGURE_ENABLE=	shell-access
 FPM_CONFIGURE_ENABLE=		fpm
@@ -67,6 +70,9 @@
 PAM_CONFIGURE_WITH=		libpam
 LDPD_CONFIGURE_ENABLE=		ldpd
 MULTIPATH_CONFIGURE_ON=		--enable-multipath=64
+RPKI_CONFIGURE_ENABLE=	rpki
+RPKI_LIB_DEPENDS=		librtr.so:net-mgmt/rtrlib \
+						libssh.so:security/libssh
 TCMALLOC_CONFIGURE_ENABLE=	tcmalloc
 TCMALLOC_LIB_DEPENDS=		libtcmalloc.so:devel/google-perftools
 SNMP_CONFIGURE_ENABLE=		snmp
Index: net/frr/distinfo
===================================================================
--- net/frr/distinfo	(revision 464399)
+++ net/frr/distinfo	(working copy)
@@ -1,3 +1,3 @@
-TIMESTAMP = 1515855735
-SHA256 (FRRouting-frr-frr-3.0.3_GH0.tar.gz) = c850d6f7ae389e27acb1e808e01e561b0db7e0e35a902ca18ce53839c7906b7b
-SIZE (FRRouting-frr-frr-3.0.3_GH0.tar.gz) = 3168397
+TIMESTAMP = 1520970456
+SHA256 (FRRouting-frr-frr-4.0_GH0.tar.gz) = a9932ef116106d56b0e17aa569aa56a458acdd50e0d07c042fd5cc725bf742cc
+SIZE (FRRouting-frr-frr-4.0_GH0.tar.gz) = 3989206
Index: net/frr/files/patch-Makefile.am
===================================================================
--- net/frr/files/patch-Makefile.am	(revision 464399)
+++ net/frr/files/patch-Makefile.am	(working copy)
@@ -1,11 +1,11 @@
---- Makefile.am.orig	2017-03-08 14:59:07 UTC
+--- Makefile.am.orig	2018-03-13 22:21:12 UTC
 +++ Makefile.am
-@@ -4,7 +4,7 @@ SUBDIRS = lib qpb fpm @ZEBRA@ @LIBRFP@ @RFPTEST@ \
- 	 @BGPD@ @RIPD@ @RIPNGD@ @OSPFD@ @OSPF6D@ @LDPD@ \
-          @ISISD@ @PIMD@ @NHRPD@ \
- 	 @WATCHFRR@ @VTYSH@ @OSPFCLIENT@ @DOC@ m4 @pkgsrcdir@ \
--         redhat @SOLARIS@ tests tools cumulus snapcraft
-+         redhat @SOLARIS@ tests cumulus snapcraft
+@@ -41,7 +41,7 @@ include zebra/subdir.am
+ include watchfrr/subdir.am
+ include qpb/subdir.am
+ include fpm/subdir.am
+-include tools/subdir.am
++#include tools/subdir.am
  
- DIST_SUBDIRS = lib qpb fpm zebra bgpd ripd ripngd ospfd ospf6d ldpd \
- 	  isisd watchfrr vtysh ospfclient doc m4 pkgsrc redhat tests \
+ include ripd/subdir.am
+ include ripngd/subdir.am
Index: net/frr/files/patch-vtysh_vtysh.c
===================================================================
--- net/frr/files/patch-vtysh_vtysh.c	(revision 464399)
+++ net/frr/files/patch-vtysh_vtysh.c	(nonexistent)
@@ -1,11 +0,0 @@
---- vtysh/vtysh.c.orig	2017-08-09 12:56:09.000000000 +0200
-+++ vtysh/vtysh.c	2017-08-22 22:04:36.068181000 +0200
-@@ -249,7 +249,7 @@
- 	if (pager_defined)
- 		vtysh_pager_name = strdup(pager_defined);
- 	else
--		vtysh_pager_name = strdup("more");
-+		vtysh_pager_name = strdup("cat");
- }
- 
- /* Command execution over the vty interface. */

Property changes on: net/frr/files/patch-vtysh_vtysh.c
___________________________________________________________________
Deleted: fbsd:nokeywords
## -1 +0,0 ##
-yes
\ No newline at end of property
Deleted: svn:eol-style
## -1 +0,0 ##
-native
\ No newline at end of property
Deleted: svn:mime-type
## -1 +0,0 ##
-text/plain
\ No newline at end of property
Index: net/frr/pkg-plist
===================================================================
--- net/frr/pkg-plist	(revision 464399)
+++ net/frr/pkg-plist	(working copy)
@@ -5,12 +5,21 @@
 include/frr/buffer.h
 include/frr/checksum.h
 include/frr/command.h
+include/frr/command_graph.h
 include/frr/command_match.h
+include/frr/compiler.h
 include/frr/csv.h
 include/frr/distribute.h
+include/frr/eigrpd/eigrp_dump.h
+include/frr/eigrpd/eigrp_topology.h
+include/frr/eigrpd/eigrpd.h
 include/frr/event_counter.h
+include/frr/ferr.h
 include/frr/fifo.h
 include/frr/filter.h
+include/frr/freebsd-queue.h
+include/frr/frr_pthread.h
+include/frr/frratomic.h
 include/frr/getopt.h
 include/frr/graph.h
 include/frr/hash.h
@@ -18,6 +27,7 @@
 include/frr/if.h
 include/frr/if_rmap.h
 include/frr/imsg.h
+include/frr/ipaddr.h
 include/frr/jhash.h
 include/frr/json.h
 include/frr/keychain.h
@@ -53,9 +63,13 @@
 include/frr/privs.h
 include/frr/ptm_lib.h
 include/frr/pw.h
+include/frr/queue.h
 include/frr/qobj.h
+include/frr/ringbuf.h
 include/frr/route_types.h
 include/frr/routemap.h
+include/frr/sbuf.h
+include/frr/sha256.h
 include/frr/sigevent.h
 include/frr/spf_backoff.h
 include/frr/skiplist.h
@@ -66,11 +80,15 @@
 include/frr/stream.h
 include/frr/systemd.h
 include/frr/table.h
+include/frr/termtable.h
 include/frr/thread.h
 include/frr/vector.h
 include/frr/version.h
+include/frr/vlan.h
+include/frr/vrf_int.h
 include/frr/vrf.h
 include/frr/vty.h
+include/frr/vxlan.h
 include/frr/wheel.h
 include/frr/workqueue.h
 include/frr/zassert.h
@@ -84,7 +102,10 @@
 lib/libfrrospfapiclient.so
 lib/libfrrospfapiclient.so.0
 lib/libfrrospfapiclient.so.0.0.0
+lib/frr/modules/bgpd_rpki.so
 man/man1/frr.1.gz
+man/man8/eigrpd.8.gz
+man/man8/frr-args.8.gz
 man/man8/isisd.8.gz
 man/man8/watchfrr.8.gz
 %%VTYSH%%man/man1/vtysh.1.gz
@@ -96,8 +117,10 @@
 man/man8/ripd.8.gz
 man/man8/ripngd.8.gz
 man/man8/zebra.8.gz
+sbin/babeld
 sbin/bgpd
 %%LDPD%%sbin/ldpd
+sbin/eigrpd
 sbin/ospf6d
 sbin/ospfclient
 sbin/ospfd
@@ -106,11 +129,12 @@
 sbin/zebra
 sbin/isisd
 sbin/rfptest
-sbin/ssd
 sbin/watchfrr
+%%ETCDIR%%/babeld.conf.sample
 %%ETCDIR%%/bgpd.conf.sample
 %%ETCDIR%%/bgpd.conf.sample2
 %%ETCDIR%%/bgpd.conf.vnc.sample
+%%ETCDIR%%/eigrpd.conf.sample
 %%ETCDIR%%/isisd.conf.sample
 %%LDPD%%%%ETCDIR%%/ldpd.conf.sample
 %%ETCDIR%%/ospf6d.conf.sample
