Index: GIDs
===================================================================
--- GIDs	(revision 444211)
+++ GIDs	(working copy)
@@ -93,7 +93,7 @@
 avenger:*:149:
 cbsd:*:150:
 # free: 151
-# free: 152
+frrvty:*:152:frr
 rbldns:*:153:
 trircd:*:154:
 xorp:*:155:
@@ -109,7 +109,7 @@
 uwsgi:*:165:
 yule:*:166:
 ceph:*:167:
-# free: 168
+frr:*:168:
 acme:*:169:
 tile38:*:170:
 sfs:*:171:
Index: UIDs
===================================================================
--- UIDs	(revision 444211)
+++ UIDs	(working copy)
@@ -114,7 +114,7 @@
 uwsgi:*:165:165::0:0:uwsgi Daemon:/nonexistent:/usr/sbin/nologin
 yule:*:166:166::0:0:Samhain Daemon:/nonexistent:/usr/sbin/nologin
 ceph:*:167:167::0:0:Ceph Daemons:/var/lib/ceph:/usr/sbin/nologin
-# free: 168
+frr:*:168:168::0:0:FRRouting user:/var/empty:/usr/sbin/nologin
 acme:*:169:169::0:0:ACME protocol client:/var/db/acme:/bin/sh
 tile38:*:170:170::0:0:Tile38 User:/var/db/tile38:/usr/sbin/nologin
 sfs:*:171:171::0:0:Self-Certifying File System:/nonexistent:/usr/sbin/nologin
Index: net/Makefile
===================================================================
--- net/Makefile	(revision 444211)
+++ net/Makefile	(working copy)
@@ -151,6 +151,7 @@
     SUBDIR += freerdp1
     SUBDIR += freeswitch
     SUBDIR += freevrrpd
+	SUBDIR += frr
     SUBDIR += fspclient
     SUBDIR += fspd
     SUBDIR += fsplib
