diff --git net-mgmt/nstat/Makefile net-mgmt/nstat/Makefile
index 9461efcd6ab0..61424a8daca0 100644
--- net-mgmt/nstat/Makefile
+++ net-mgmt/nstat/Makefile
@@ -11,8 +11,8 @@ LICENSE=	BSD2CLAUSE
 LICENSE_FILE=	${WRKSRC}/LICENSE
 
 RUN_DEPENDS=	${RUN_DEPENDS_${ARCH}}
-RUN_DEPENDS_i386=	pcm.x:sysutils/intel-pcm
-RUN_DEPENDS_amd64=	pcm.x:sysutils/intel-pcm
+RUN_DEPENDS_i386=	pcm:sysutils/intel-pcm
+RUN_DEPENDS_amd64=	pcm:sysutils/intel-pcm
 
 USE_GITHUB=	yes
 GH_ACCOUNT=	gallatin
