Index: net/frr7/Makefile
===================================================================
--- net/frr7/Makefile	(revision 536439)
+++ net/frr7/Makefile	(working copy)
@@ -1,10 +1,10 @@
 # $FreeBSD$
 
 PORTNAME=	frr
-PORTVERSION=	7.3
+PORTVERSION=	7.3.1
 DISTVERSIONPREFIX=	frr-
-PORTREVISION=	2
 CATEGORIES=	net
+CATEGORIES=	net
 PKGNAMESUFFIX=	7
 
 MAINTAINER=	olivier@FreeBSD.org
Index: net/frr7/distinfo
===================================================================
--- net/frr7/distinfo	(revision 536439)
+++ net/frr7/distinfo	(working copy)
@@ -1,3 +1,3 @@
-TIMESTAMP = 1582712493
-SHA256 (FRRouting-frr-frr-7.3_GH0.tar.gz) = c209ed66a0331221649b8c21dc8c6704ce1696e2017ef4d48f8c2ac6b1fa01b7
-SIZE (FRRouting-frr-frr-7.3_GH0.tar.gz) = 5782842
+TIMESTAMP = 1589228942
+SHA256 (FRRouting-frr-frr-7.3.1_GH0.tar.gz) = 81129f83ce9ae31c983bb02b3d8e3e457604bab46f01411c945d9c2cfef1f5c8
+SIZE (FRRouting-frr-frr-7.3.1_GH0.tar.gz) = 5785833
Index: net/frr7/files/patch-bgpd_bgp__route.c
===================================================================
--- net/frr7/files/patch-bgpd_bgp__route.c	(revision 536439)
+++ net/frr7/files/patch-bgpd_bgp__route.c	(nonexistent)
@@ -1,29 +0,0 @@
---- bgpd/bgp_route.c.orig	2020-02-14 14:03:09 UTC
-+++ bgpd/bgp_route.c
-@@ -9922,7 +9922,7 @@ static int bgp_show_route_in_table(struct vty *vty, st
- 		}
- 	} else if (safi == SAFI_EVPN) {
- 		struct bgp_node *longest_pfx;
--		bool is_exact_pfxlen_match = FALSE;
-+		bool is_exact_pfxlen_match = false;
- 
- 		for (rn = bgp_table_top(rib); rn; rn = bgp_route_next(rn)) {
- 			if (prd && memcmp(rn->p.u.val, prd->val, 8) != 0)
-@@ -9932,7 +9932,7 @@ static int bgp_show_route_in_table(struct vty *vty, st
- 				continue;
- 
- 			longest_pfx = NULL;
--			is_exact_pfxlen_match = FALSE;
-+			is_exact_pfxlen_match = false;
- 			/*
- 			 * Search through all the prefixes for a match.  The
- 			 * pfx's are enumerated in ascending order of pfxlens.
-@@ -9951,7 +9951,7 @@ static int bgp_show_route_in_table(struct vty *vty, st
- 					int type5_pfxlen =
- 					   bgp_evpn_get_type5_prefixlen(&rm->p);
- 					if (type5_pfxlen == match.prefixlen) {
--						is_exact_pfxlen_match = TRUE;
-+						is_exact_pfxlen_match = true;
- 						bgp_unlock_node(rm);
- 						break;
- 					}

Property changes on: net/frr7/files/patch-bgpd_bgp__route.c
___________________________________________________________________
Deleted: fbsd:nokeywords
## -1 +0,0 ##
-yes
\ No newline at end of property
Deleted: svn:eol-style
## -1 +0,0 ##
-native
\ No newline at end of property
Deleted: svn:mime-type
## -1 +0,0 ##
-text/plain
\ No newline at end of property
