Index: net/bird/files/patch-sysdep-bsd-sysio.h
===================================================================
--- net/bird/files/patch-sysdep-bsd-sysio.h	(revision 0)
+++ net/bird/files/patch-sysdep-bsd-sysio.h	(working copy)
@@ -0,0 +1,22 @@
+--- sysdep/bsd/sysio.h.orig	2014-09-26 09:31:32.284787124 +0200
++++ sysdep/bsd/sysio.h	2014-09-26 09:31:35.391786781 +0200
+@@ -6,6 +6,8 @@
+  *	Can be freely distributed and used under the terms of the GNU GPL.
+  */
+ 
++#include <sys/param.h>
++
+ #include <net/if_dl.h>
+ #include <netinet/in_systm.h> // Workaround for some BSDs
+ #include <netinet/ip.h>
+@@ -175,8 +179,8 @@
+   ip->ip_src = ipa_to_in4(s->saddr);
+   ip->ip_dst = ipa_to_in4(s->daddr);
+ 
+-#ifdef __OpenBSD__
+-  /* OpenBSD expects ip_len in network order, other BSDs expect host order */
++#if defined __OpenBSD__ || (defined __FreeBSD__ && (__FreeBSD_version >= 1100030))
++  /* OpenBSD and FreeBSD (since 11) expects ip_len in network order, other BSDs expect host order */
+   ip->ip_len = htons(ip->ip_len);
+ #endif
+ }

Property changes on: net/bird/files/patch-sysdep-bsd-sysio.h
___________________________________________________________________
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
