Index: net-mgmt/bsnmp-ucd/Makefile
===================================================================
--- net-mgmt/bsnmp-ucd/Makefile	(revision 528340)
+++ net-mgmt/bsnmp-ucd/Makefile	(working copy)
@@ -3,6 +3,7 @@
 
 PORTNAME=	bsnmp-ucd
 PORTVERSION=	0.4.4
+PORTREVISION=	1
 CATEGORIES=	net-mgmt
 
 MAINTAINER=	trociny@FreeBSD.org
@@ -10,10 +11,11 @@
 
 LICENSE=	BSD2CLAUSE
 
+USES=		uidfix
+
 USE_GITHUB=	yes
 GH_ACCOUNT=	trociny
 
-USES=		uidfix
 MAKE_ENV+=	NO_OBJ=yes
 
 OPTIONS_DEFINE=	EXAMPLES
Index: net-mgmt/bsnmp-ucd/files/patch-mibdio.c
===================================================================
--- net-mgmt/bsnmp-ucd/files/patch-mibdio.c	(nonexistent)
+++ net-mgmt/bsnmp-ucd/files/patch-mibdio.c	(working copy)
@@ -0,0 +1,11 @@
+--- mibdio.c.orig	2020-03-13 16:46:30 UTC
++++ mibdio.c
+@@ -172,7 +172,7 @@ update_dio_data(void *arg __unused)
+ 		if (diop->_busy_time.sec > 0) {
+ 			busy_time = dev.busy_time.sec - diop->_busy_time.sec +
+ 			    (double)(dev.busy_time.frac - diop->_busy_time.frac)
+-			    / 0xffffffffffffffffull;
++			    / (double)0xffffffffffffffffull;
+ 			if (busy_time < 0) /* FP loss of precision near zero. */
+ 				busy_time = 0;
+ 			percent = busy_time * 100 / interval;

Property changes on: net-mgmt/bsnmp-ucd/files/patch-mibdio.c
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
