diff --git Mk/bsd.port.mk Mk/bsd.port.mk
index 13785ff0f234..4410e25a8c87 100644
--- Mk/bsd.port.mk
+++ Mk/bsd.port.mk
@@ -2061,6 +2061,8 @@ MAKE_ENV+=		PREFIX=${PREFIX} \
 			CXX="${CXX}" CXXFLAGS="${CXXFLAGS}" \
 			MANPREFIX="${MANPREFIX}"
 
+MAKE_ENV+=	TARGET_ARCH=
+
 # Add -fno-strict-aliasing to CFLAGS with optimization level -O2 or higher.
 # gcc 4.x enable strict aliasing optimization with -O2 which is known to break
 # a lot of ports.
