Index: net/pim6-tools/Makefile
===================================================================
--- net/pim6-tools/Makefile	(revision 516926)
+++ net/pim6-tools/Makefile	(working copy)
@@ -3,7 +3,7 @@
 PORTNAME=	pim6
 PORTVERSION=	20061214
 CATEGORIES=	net
-MASTER_SITES=	SF/mcast-tools
+MASTER_SITES=	http://dev.bsdrp.net/FreeBSD/
 PKGNAMESUFFIX=	-tools
 DISTNAME=	mcast-tools-${PORTVERSION}
 
@@ -10,8 +10,6 @@
 MAINTAINER=	hrs@FreeBSD.org
 COMMENT=	IPv6 multicast tools
 
-BROKEN=		unfetchable
-
 LICENSE=	BSD3CLAUSE
 LICENSE_FILE=	${WRKSRC}/COPYING
 
