diff --git net/pimd/Makefile net/pimd/Makefile
index f4e8c572d5af..6f9f24fac619 100644
--- net/pimd/Makefile
+++ net/pimd/Makefile
@@ -1,6 +1,5 @@
 PORTNAME=	pimd
-PORTVERSION=	2.3.2
-PORTREVISION=	1
+PORTVERSION=	2.3.2b
 CATEGORIES=	net
 
 MAINTAINER=	olivier@FreeBSD.org
@@ -10,23 +9,33 @@ WWW=		http://troglobit.com/pimd.html
 LICENSE=	BSD3CLAUSE
 LICENSE_FILE=	${WRKSRC}/LICENSE
 
+USES=		autoreconf:build
+
 USE_GITHUB=	yes
 GH_ACCOUNT=	troglobit
 GH_PROJECT=	libite:libite
-GH_TAGNAME=	v1.4.4:libite
+GH_TAGNAME=	17ae62f\
+		v1.4.4:libite
 GH_SUBDIR=	libite:libite
 
 USE_RC_SUBR=	${PORTNAME}
 
-HAS_CONFIGURE=	yes
+GNU_CONFIGURE=	yes
+
 MAKE_ARGS=	prefix="${PREFIX}" sysconfdir="${PREFIX}/etc/"
 
+pre-configure:
+	cd ${WRKSRC} && ./autogen.sh
+
 do-install:
-	${INSTALL_PROGRAM} ${WRKSRC}/pimd ${STAGEDIR}${PREFIX}/sbin/
+	${INSTALL_PROGRAM} ${WRKSRC}/src/pimd ${STAGEDIR}${PREFIX}/sbin/
+	${INSTALL_PROGRAM} ${WRKSRC}/src/pimctl ${STAGEDIR}${PREFIX}/sbin/
 	${INSTALL_DATA} ${WRKSRC}/pimd.conf \
 		${STAGEDIR}${PREFIX}/etc/pimd.conf.sample
 	${INSTALL_SCRIPT} ${FILESDIR}/pimd.in \
 		${STAGEDIR}${PREFIX}/etc/rc.d/pimd
-	${INSTALL_MAN} ${WRKSRC}/pimd.8 ${STAGEDIR}${PREFIX}/share/man/man8/
+	${INSTALL_MAN} ${WRKSRC}/man/pimd.8 ${STAGEDIR}${PREFIX}/share/man/man8/
+	${INSTALL_MAN} ${WRKSRC}/man/pimctl.8 ${STAGEDIR}${PREFIX}/share/man/man8/
+	${INSTALL_MAN} ${WRKSRC}/man/pimd.conf.5 ${STAGEDIR}${PREFIX}/share/man/man5/
 
 .include <bsd.port.mk>
diff --git net/pimd/distinfo net/pimd/distinfo
index 04f5043941d3..6d6ed53b9fbe 100644
--- net/pimd/distinfo
+++ net/pimd/distinfo
@@ -1,4 +1,5 @@
-SHA256 (troglobit-pimd-2.3.2_GH0.tar.gz) = b6cda75a55952fac06fdc50a95ab7ee2f94f8568de365c49c49a2acf078488eb
-SIZE (troglobit-pimd-2.3.2_GH0.tar.gz) = 237277
+TIMESTAMP = 1733871538
+SHA256 (troglobit-pimd-2.3.2b-17ae62f_GH0.tar.gz) = d247e190eedae2314fe5e442fe79f0d9c4260ee62077fa38827d5e53c90eac2c
+SIZE (troglobit-pimd-2.3.2b-17ae62f_GH0.tar.gz) = 1390204
 SHA256 (troglobit-libite-v1.4.4_GH0.tar.gz) = c4327e7d3ade7efb191d02a4f9ffaf7b8e86e7e7f91a2f7e8b600e2102330607
 SIZE (troglobit-libite-v1.4.4_GH0.tar.gz) = 36086
diff --git net/pimd/files/pimd.in net/pimd/files/pimd.in
index d3339b5e9c70..251bb9ccd71f 100644
--- net/pimd/files/pimd.in
+++ net/pimd/files/pimd.in
@@ -2,7 +2,7 @@
 #
 # PROVIDE: pimd
 # REQUIRE: netif routing
-# KEYWORD: nojail
+# KEYWORD: nojailvnet
 
 #
 # Add the following to /etc/rc.conf[.local] to enable this service
@@ -17,6 +17,7 @@ rcvar="${name}_enable"
 command="%%PREFIX%%/sbin/${name}"
 required_files="%%PREFIX%%/etc/${name}.conf"
 extra_commands=reload
+required_modules="ip_mroute
 
 load_rc_config $name
 : ${pimd_enable="NO"}
diff --git net/pimd/pkg-plist net/pimd/pkg-plist
index 723d3999ad75..18039f16831c 100644
--- net/pimd/pkg-plist
+++ net/pimd/pkg-plist
@@ -1,3 +1,6 @@
 @sample etc/pimd.conf.sample
 sbin/pimd
+sbin/pimctl
 share/man/man8/pimd.8.gz
+share/man/man8/pimctl.8.gz
+share/man/man5/pimd.conf.5.gz
