Index: net/pimdd/Makefile
===================================================================
--- net/pimdd/Makefile	(revision 444518)
+++ net/pimdd/Makefile	(working copy)
@@ -3,7 +3,7 @@
 
 PORTNAME=	pimdd
 PORTVERSION=	0.2.1.0
-PORTREVISION=	2
+PORTREVISION=	3
 CATEGORIES=	net
 MASTER_SITES=	http://antc.uoregon.edu/PIMDM/
 DISTNAME=	pimd-dense
Index: net/pimdd/files/patch-Makefile
===================================================================
--- net/pimdd/files/patch-Makefile	(revision 444518)
+++ net/pimdd/files/patch-Makefile	(working copy)
@@ -17,7 +17,7 @@
  ### Compilation flags for different platforms. Uncomment only one of them
  ## FreeBSD
 -CFLAGS= -Wall -g	-Iinclude/freebsd ${MCAST_INCLUDE} ${SNMPDEF} ${RSRRDEF} ${MISCDEFS} -DFreeBSD -DPIM
-+CFLAGS+=	${MCAST_INCLUDE} ${SNMPDEF} ${RSRRDEF} ${MISCDEFS} -DFreeBSD -DPIM -Dlog=logit ${CPPFLAGS}
++CFLAGS+=	${MCAST_INCLUDE} ${SNMPDEF} ${RSRRDEF} ${MISCDEFS} -DFreeBSD -DPIM -Dlog=logit ${CPPFLAGS} -DRAW_INPUT_IS_RAW -DRAW_OUTPUT_IS_RAW
  
  ## NetBSD   -DNetBSD is done by OS
  #CFLAGS= -Wall -g	-Iinclude/netbsd ${MCAST_INCLUDE} ${SNMPDEF} ${RSRRDEF} ${MISCDEFS} -DPIM
Index: net/pimdd/files/pimdd.in
===================================================================
--- net/pimdd/files/pimdd.in	(revision 444518)
+++ net/pimdd/files/pimdd.in	(working copy)
@@ -18,6 +18,7 @@
 command="%%PREFIX%%/sbin/${name}"
 required_files="%%PREFIX%%/etc/${name}.conf"
 extra_commands=reload
+required_modules="ip_mroute"
 
 load_rc_config $name
 : ${pimdd_enable="NO"}
