diff --git a/net/frr8/Makefile b/net/frr8/Makefile
index 0e799fbedf57..43c887ac9d93 100644
--- a/net/frr8/Makefile
+++ b/net/frr8/Makefile
@@ -1,5 +1,5 @@
 PORTNAME=	frr
-PORTVERSION=	8.3
+PORTVERSION=	8.3.1
 DISTVERSIONPREFIX=	frr-
 CATEGORIES=	net
 .if defined(PYTHONTOOLS)
diff --git a/net/frr8/distinfo b/net/frr8/distinfo
index 06782330e377..40b5d54cd5ca 100644
--- a/net/frr8/distinfo
+++ b/net/frr8/distinfo
@@ -1,3 +1,3 @@
-TIMESTAMP = 1658433791
-SHA256 (FRRouting-frr-frr-8.3_GH0.tar.gz) = bfc625936c72e64397acfa9182b1977a86b41ff4b3be96b41a59a6cb1621f4a6
-SIZE (FRRouting-frr-frr-8.3_GH0.tar.gz) = 9676363
+TIMESTAMP = 1662385513
+SHA256 (FRRouting-frr-frr-8.3.1_GH0.tar.gz) = 9b0430d01f8da2d780ac5e41de2cf4e53abff0ed06a0cc75d745f4b1f29e5b9e
+SIZE (FRRouting-frr-frr-8.3.1_GH0.tar.gz) = 9676895
