Index: print/texinfo/files/patch-tp-Texinfo-XS-parsetexi-api.c
===================================================================
--- print/texinfo/files/patch-tp-Texinfo-XS-parsetexi-api.c	(nonexistent)
+++ print/texinfo/files/patch-tp-Texinfo-XS-parsetexi-api.c	(working copy)
@@ -0,0 +1,12 @@
+--- tp/Texinfo/XS/parsetexi/api.c.orig	2019-08-25 17:11:45 UTC
++++ tp/Texinfo/XS/parsetexi/api.c
+@@ -25,7 +25,9 @@
+ 
+ #undef context
+ 
++#if defined(HAVE_LIBINTL_H)
+ #include <libintl.h>
++#endif
+ 
+ #include <stdlib.h>
+ #include <stdio.h>

Property changes on: print/texinfo/files/patch-tp-Texinfo-XS-parsetexi-api.c
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: print/texinfo/files/patch-tp-Texinfo-XS-parsetexi-errors.c
===================================================================
--- print/texinfo/files/patch-tp-Texinfo-XS-parsetexi-errors.c	(nonexistent)
+++ print/texinfo/files/patch-tp-Texinfo-XS-parsetexi-errors.c	(working copy)
@@ -0,0 +1,12 @@
+--- tp/Texinfo/XS/parsetexi/errors.c.orig	2019-08-25 17:11:45 UTC
++++ tp/Texinfo/XS/parsetexi/errors.c
+@@ -14,7 +14,9 @@
+    along with this program.  If not, see <http://www.gnu.org/licenses/>. */
+ 
+ #include <config.h>
++#if defined(HAVE_LIBINTL_H)
+ #include <libintl.h>
++#endif
+ 
+ #include <stdlib.h>
+ #include <stdarg.h>

Property changes on: print/texinfo/files/patch-tp-Texinfo-XS-parsetexi-errors.c
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
