diff --git net/libyang3/Makefile net/libyang3/Makefile
index f9d665477ad4..f887391f3bda 100644
--- net/libyang3/Makefile
+++ net/libyang3/Makefile
@@ -14,7 +14,7 @@ LICENSE_FILE=	${WRKSRC}/LICENSE
 LIB_DEPENDS=	libpcre2-8.so:devel/pcre2
 TEST_DEPENDS=	cmocka>0:sysutils/cmocka
 
-USES=	cmake compiler:c++11-lang pkgconfig
+USES=	cmake compiler:c++11-lang localbase pkgconfig
 USE_LDCONFIG=	yes
 
 USE_GITHUB=	yes
