Index: net/pimd/Makefile
===================================================================
--- net/pimd/Makefile	(revision 394578)
+++ net/pimd/Makefile	(working copy)
@@ -1,8 +1,9 @@
 # $FreeBSD$
 
 PORTNAME=	pimd
-PORTVERSION=	2.2.0
+PORTVERSION=	2.3.0
 CATEGORIES=	net
+MASTER_SITES=	http://ftp.troglobit.com/pimd/
 
 MAINTAINER=	olivier@cochard.me
 COMMENT=	Lightweight stand-alone PIM-SM v2 multicast routing daemon
@@ -9,10 +10,6 @@
 
 LICENSE=	BSD3CLAUSE
 
-USE_GITHUB=	yes
-GH_ACCOUNT=	troglobit
-
-USES=		tar:bzip2
 HAS_CONFIGURE=	yes
 
 post-patch:
Index: net/pimd/distinfo
===================================================================
--- net/pimd/distinfo	(revision 394578)
+++ net/pimd/distinfo	(working copy)
@@ -1,2 +1,2 @@
-SHA256 (troglobit-pimd-2.2.0_GH0.tar.bz2) = a6d969e235323852afdd1c9aebc15a2ff1e4da75e8277a736b3be9b2b484e30e
-SIZE (troglobit-pimd-2.2.0_GH0.tar.bz2) = 223012
+SHA256 (pimd-2.3.0.tar.gz) = 9f1755b718a359b6bcde46a6f98cb84ff9fcbdc98ad9612bfe204e999d77b849
+SIZE (pimd-2.3.0.tar.gz) = 263088
