Index: net/frr/Makefile
===================================================================
--- net/frr/Makefile	(revision 458934)
+++ net/frr/Makefile	(working copy)
@@ -1,7 +1,7 @@
 # $FreeBSD$
 
 PORTNAME=	frr
-PORTVERSION=	3.0.2
+PORTVERSION=	3.0.3
 DISTVERSIONPREFIX=	frr-
 PORTREVISION=	1
 CATEGORIES=	net ipv6
Index: net/frr/distinfo
===================================================================
--- net/frr/distinfo	(revision 458934)
+++ net/frr/distinfo	(working copy)
@@ -1,3 +1,3 @@
-TIMESTAMP = 1510233191
-SHA256 (FRRouting-frr-frr-3.0.2_GH0.tar.gz) = b070a5175ec322ba8661ea01a416d643e794739e2e4281e6ee4d3d9a49199812
-SIZE (FRRouting-frr-frr-3.0.2_GH0.tar.gz) = 3167879
+TIMESTAMP = 1515855735
+SHA256 (FRRouting-frr-frr-3.0.3_GH0.tar.gz) = c850d6f7ae389e27acb1e808e01e561b0db7e0e35a902ca18ce53839c7906b7b
+SIZE (FRRouting-frr-frr-3.0.3_GH0.tar.gz) = 3168397
