diff --git sysutils/dtrace-toolkit/Makefile sysutils/dtrace-toolkit/Makefile
index 156c31a4068c..8157bfd41122 100644
--- sysutils/dtrace-toolkit/Makefile
+++ sysutils/dtrace-toolkit/Makefile
@@ -12,7 +12,7 @@ LICENSE_FILE=	${WRKSRC}/Docs/cddl1.txt
 
 RUN_DEPENDS=	ksh93:shells/ksh
 
-USES=		perl5 python ruby shebangfix
+USES=		perl5 python shebangfix
 USE_GITHUB=	yes
 GH_ACCOUNT=	opendtrace
 GH_PROJECT=	toolkit
