Index: net/bird2/Makefile
===================================================================
--- net/bird2/Makefile	(revision 551493)
+++ net/bird2/Makefile	(working copy)
@@ -3,7 +3,7 @@
 
 PORTNAME=	bird
 DISTVERSION=	2.0.7
-PORTREVISION=	1
+PORTREVISION=	2
 CATEGORIES=	net
 MASTER_SITES=	ftp://bird.network.cz/pub/bird/
 PKGNAMESUFFIX=	2
Index: net/bird2/files/patch-sysdep_bsd_sysio.h
===================================================================
--- net/bird2/files/patch-sysdep_bsd_sysio.h	(nonexistent)
+++ net/bird2/files/patch-sysdep_bsd_sysio.h	(working copy)
@@ -0,0 +1,14 @@
+--- sysdep/bsd/sysio.h.orig	2020-10-06 08:11:53 UTC
++++ sysdep/bsd/sysio.h
+@@ -12,11 +12,6 @@
+ #include <sys/param.h>
+ 
+ 
+-#ifdef __FreeBSD__
+-/* Should be defined in sysdep/cf/bsd.h, but it is flavor-specific */
+-#define CONFIG_DONTROUTE_UNICAST
+-#endif
+-
+ #ifdef __NetBSD__
+ 
+ #ifndef IP_RECVTTL

Property changes on: net/bird2/files/patch-sysdep_bsd_sysio.h
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
