diff -ruN sysutils/ipmitool.orig/files/patch-ipmi_chassis.c sysutils/ipmitool/files/patch-ipmi_chassis.c
--- sysutils/ipmitool.orig/files/patch-ipmi_chassis.c	1970-01-01 00:00:00.000000000 +0000
+++ sysutils/ipmitool/files/patch-ipmi_chassis.c	2012-10-15 07:40:40.000000000 +0000
@@ -0,0 +1,10 @@
+--- lib/ipmi_chassis.c.orig	2012-10-15 07:38:19.000000000 +0000
++++ lib/ipmi_chassis.c	2012-10-15 07:38:46.000000000 +0000
+@@ -41,6 +41,7 @@
+ #include <ipmitool/ipmi_intf.h>
+ #include <ipmitool/ipmi_strings.h>
+ #include <ipmitool/ipmi_chassis.h>
++#include <ipmitool/bswap.h>
+ 
+ extern int verbose;
+ 
diff -ruN sysutils/ipmitool.orig/files/patch-ipmi_pef.c sysutils/ipmitool/files/patch-ipmi_pef.c
--- sysutils/ipmitool.orig/files/patch-ipmi_pef.c	1970-01-01 00:00:00.000000000 +0000
+++ sysutils/ipmitool/files/patch-ipmi_pef.c	2012-10-15 07:41:58.000000000 +0000
@@ -0,0 +1,10 @@
+--- lib/ipmi_pef.c.orig	2012-10-15 07:41:07.000000000 +0000
++++ lib/ipmi_pef.c	2012-10-15 07:41:36.000000000 +0000
+@@ -39,6 +39,7 @@
+ #include <ipmitool/ipmi.h>
+ #include <ipmitool/ipmi_intf.h>
+ #include <ipmitool/ipmi_pef.h>
++#include <ipmitool/bswap.h>
+ 
+ extern int verbose;
+ /*
