Index: net/pimdd/Makefile
===================================================================
--- net/pimdd/Makefile	(revision 340640)
+++ net/pimdd/Makefile	(working copy)
@@ -3,6 +3,7 @@
 
 PORTNAME=	pimdd
 PORTVERSION=	0.2.1.0
+PORTREVISION=	1
 CATEGORIES=	net
 MASTER_SITES=	http://antc.uoregon.edu/PIMDM/
 DISTNAME=	pimd-dense
@@ -11,7 +12,6 @@
 COMMENT=	UO Dense Protocol-Independent Multicast (PIM-DM) daemon for IPv4
 
 MAKE_ARGS+=	CC="${CC}"
-USE_GCC=	any
 
 NO_MAN=	yes
 
@@ -20,7 +20,6 @@
 
 PORTDOCS=	pimd-dense.html README CHANGES RELEASE.NOTES
 
-NO_STAGE=	yes
 .include <bsd.port.pre.mk>
 
 EXTRA_PATCHES=	${FILESDIR}/extra-patch-defs.h
@@ -29,12 +28,12 @@
 	${REINPLACE_CMD} -e 's|/etc/pimd.conf|${PREFIX}/etc/pimd.conf|g' ${WRKSRC}/pathnames.h
 
 do-install:
-	${INSTALL_PROGRAM} ${WRKSRC}/pimdd ${PREFIX}/sbin/pimdd
-	${INSTALL_DATA} ${WRKSRC}/pimdd.conf ${PREFIX}/etc/pimdd.conf.sample
+	${INSTALL_PROGRAM} ${WRKSRC}/pimdd ${STAGEDIR}${PREFIX}/sbin/pimdd
+	${INSTALL_DATA} ${WRKSRC}/pimdd.conf ${STAGEDIR}${PREFIX}/etc/pimdd.conf.sample
 .if !defined(NOPORTDOCS)
-	${MKDIR} ${DOCSDIR}
+	${MKDIR} ${STAGEDIR}${DOCSDIR}
 .for i in ${PORTDOCS}
-	${INSTALL_MAN} ${WRKSRC}/${i} ${DOCSDIR}/${i}
+	${INSTALL_MAN} ${WRKSRC}/${i} ${STAGEDIR}${DOCSDIR}/${i}
 .endfor
 .endif
 
Index: net/pimdd/files/patch-Makefile
===================================================================
--- net/pimdd/files/patch-Makefile	(revision 340640)
+++ net/pimdd/files/patch-Makefile	(working copy)
@@ -1,5 +1,14 @@
---- Makefile.orig	Fri Sep 29 11:53:31 2006
-+++ Makefile	Fri Sep 29 11:53:50 2006
+--- Makefile.orig	1999-11-30 18:58:53.000000000 +0100
++++ Makefile	2013-09-08 09:39:18.947517833 +0200
+@@ -53,7 +53,7 @@
+ # Support for Resource Reservations), currently used by RSVP.
+ RSRRDEF=	-DRSRR
+ 
+-CC =		gcc
++#CC =		gcc
+ MCAST_INCLUDE=	-Iinclude
+ LDFLAGS=
+ 
 @@ -63,7 +63,7 @@
  #CONFIGCONFIGCONFIG
  ### Compilation flags for different platforms. Uncomment only one of them
Index: net/pimdd/files/patch-igmp.c
===================================================================
--- net/pimdd/files/patch-igmp.c	(revision 0)
+++ net/pimdd/files/patch-igmp.c	(working copy)
@@ -0,0 +1,14 @@
+--- igmp.c.orig	2014-01-22 00:05:49.560452946 +0100
++++ igmp.c	2014-01-22 00:06:57.098508699 +0100
+@@ -168,7 +168,11 @@
+ #ifdef RAW_INPUT_IS_RAW
+     ipdatalen = ntohs(ip->ip_len) - iphdrlen;
+ #else
++ #if __FreeBSD_version >= 1000000
++	ipdatalen = ip->ip_len - iphdrlen;
++ #else
+     ipdatalen = ip->ip_len;
++ #endif
+ #endif
+     if (iphdrlen + ipdatalen != recvlen) {
+ 	log(LOG_WARNING, 0,

Property changes on: net/pimdd/files/patch-igmp.c
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
