Index: net/bird/files/bird.in
===================================================================
--- net/bird/files/bird.in	(revision 432209)
+++ net/bird/files/bird.in	(working copy)
@@ -29,6 +29,6 @@
 : ${bird_enable="NO"}
 : ${bird_config="%%PREFIX%%/etc/bird.conf"}
 
-command_args="-c $bird_config"
+command_args="-P /var/run/${name}.pid -c $bird_config"
 
 run_rc_command "$1"
Index: net/bird/files/bird6.in
===================================================================
--- net/bird/files/bird6.in	(revision 432209)
+++ net/bird/files/bird6.in	(working copy)
@@ -29,6 +29,6 @@
 : ${bird6_enable="NO"}
 : ${bird6_config="%%PREFIX%%/etc/bird6.conf"}
 
-command_args="-c $bird6_config"
+command_args="-P /var/run/${name}.pid -c $bird6_config"
 
 run_rc_command "$1"
Index: net/bird/files/patch-sysdep-bsd-sysio.h
===================================================================
--- net/bird/files/patch-sysdep-bsd-sysio.h	(nonexistent)
+++ net/bird/files/patch-sysdep-bsd-sysio.h	(working copy)
@@ -0,0 +1,23 @@
+diff --git sysdep/bsd/sysio.h sysdep/bsd/sysio.h
+index 2610a47..9b10e6e 100644
+--- sysdep/bsd/sysio.h
++++ sysdep/bsd/sysio.h
+@@ -9,6 +9,7 @@
+ #include <net/if_dl.h>
+ #include <netinet/in_systm.h> // Workaround for some BSDs
+ #include <netinet/ip.h>
++#include <sys/param.h>
+ 
+ 
+ #ifdef __NetBSD__
+@@ -179,8 +180,8 @@ sk_prepare_ip_header(sock *s, void *hdr, int dlen)
+   ip->ip_src = ipa_to_in4(s->saddr);
+   ip->ip_dst = ipa_to_in4(s->daddr);
+ 
+-#ifdef __OpenBSD__
+-  /* OpenBSD expects ip_len in network order, other BSDs expect host order */
++#if (defined __OpenBSD__) || (defined __DragonFly__) || (defined __FreeBSD__ && (__FreeBSD_version >= 1100030))
++  /* Different BSDs have different expectations of ip_len endianity */
+   ip->ip_len = htons(ip->ip_len);
+ #endif
+ }

Property changes on: net/bird/files/patch-sysdep-bsd-sysio.h
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: net/bird/files/sysdep-bsd-sysio.h
===================================================================
--- net/bird/files/sysdep-bsd-sysio.h	(revision 432209)
+++ net/bird/files/sysdep-bsd-sysio.h	(nonexistent)
@@ -1,23 +0,0 @@
-diff --git sysdep/bsd/sysio.h sysdep/bsd/sysio.h
-index 2610a47..9b10e6e 100644
---- sysdep/bsd/sysio.h
-+++ sysdep/bsd/sysio.h
-@@ -9,6 +9,7 @@
- #include <net/if_dl.h>
- #include <netinet/in_systm.h> // Workaround for some BSDs
- #include <netinet/ip.h>
-+#include <sys/param.h>
- 
- 
- #ifdef __NetBSD__
-@@ -179,8 +180,8 @@ sk_prepare_ip_header(sock *s, void *hdr, int dlen)
-   ip->ip_src = ipa_to_in4(s->saddr);
-   ip->ip_dst = ipa_to_in4(s->daddr);
- 
--#ifdef __OpenBSD__
--  /* OpenBSD expects ip_len in network order, other BSDs expect host order */
-+#if (defined __OpenBSD__) || (defined __DragonFly__) || (defined __FreeBSD__ && (__FreeBSD_version >= 1100030))
-+  /* Different BSDs have different expectations of ip_len endianity */
-   ip->ip_len = htons(ip->ip_len);
- #endif
- }

Property changes on: net/bird/files/sysdep-bsd-sysio.h
___________________________________________________________________
Deleted: fbsd:nokeywords
## -1 +0,0 ##
-yes
\ No newline at end of property
Deleted: svn:eol-style
## -1 +0,0 ##
-native
\ No newline at end of property
Deleted: svn:mime-type
## -1 +0,0 ##
-text/plain
\ No newline at end of property
