diff --git a/sysutils/mstflint/Makefile b/sysutils/mstflint/Makefile
index 9fd6dcab207e..86e77583153e 100644
--- a/sysutils/mstflint/Makefile
+++ b/sysutils/mstflint/Makefile
@@ -49,7 +49,7 @@ USES+=	ssl
 .if ${PORT_OPTIONS:MADAB}
 CONFIGURE_ARGS+=	--enable-adb-generic-tools
 .if ${PORT_OPTIONS:MSTATIC}
-BROKEN=		does not link: expat2 no longer provides static library by default
+#BROKEN=		does not link: expat2 no longer provides static library by default
 BUILD_DEPENDS+=	xmlwf:textproc/expat2 \
 		boost-libs>0:devel/boost-libs
 .else
