Index: sysutils/intel-pcm/Makefile
===================================================================
--- sysutils/intel-pcm/Makefile	(revision 502233)
+++ sysutils/intel-pcm/Makefile	(working copy)
@@ -2,7 +2,7 @@
 # $FreeBSD$
 
 PORTNAME=	intel-pcm
-PORTVERSION=	201812.3
+PORTVERSION=	201902
 CATEGORIES=	sysutils
 
 MAINTAINER=	imp@FreeBSD.org
Index: sysutils/intel-pcm/distinfo
===================================================================
--- sysutils/intel-pcm/distinfo	(revision 502233)
+++ sysutils/intel-pcm/distinfo	(working copy)
@@ -1,3 +1,3 @@
-TIMESTAMP = 1545499065
-SHA256 (opcm-pcm-201812.3-201710_GH0.tar.gz) = 772c7b6310ba9c89931ff6108cc71fa0cbe20540ab6a53503f5f61594c9f661d
-SIZE (opcm-pcm-201812.3-201710_GH0.tar.gz) = 1103994
+TIMESTAMP = 1558465693
+SHA256 (opcm-pcm-201902-201710_GH0.tar.gz) = 772c7b6310ba9c89931ff6108cc71fa0cbe20540ab6a53503f5f61594c9f661d
+SIZE (opcm-pcm-201902-201710_GH0.tar.gz) = 1103994
