Index: net/dhcprelya/Makefile
===================================================================
--- net/dhcprelya/Makefile	(revision 419803)
+++ net/dhcprelya/Makefile	(working copy)
@@ -3,6 +3,7 @@
 
 PORTNAME=	dhcprelya
 PORTVERSION=	4.9
+PORTREVISION=	1
 CATEGORIES=	net
 
 MAINTAINER=	sem@FreeBSD.org
Index: net/dhcprelya/files/patch-dhcprelya.c
===================================================================
--- net/dhcprelya/files/patch-dhcprelya.c	(nonexistent)
+++ net/dhcprelya/files/patch-dhcprelya.c	(working copy)
@@ -0,0 +1,13 @@
+diff --git dhcprelya.c dhcprelya.c
+index 1b973aa..1788122 100644
+--- dhcprelya.c
++++ dhcprelya.c
+@@ -401,7 +401,7 @@ process_server_answer()
+ 	}
+ 
+ 	timeout.tv_sec = 0;
+-	timeout.tv_usec = 10;
++	timeout.tv_usec = 100;
+ 	if (select(fdmax + 1, &fds, NULL, NULL, &timeout)) {
+ 		for (i = 0; i < if_num; i++)
+ 			if (FD_ISSET(ifs[i]->fd, &fds)) {

Property changes on: net/dhcprelya/files/patch-dhcprelya.c
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
