Index: devel/scons/Makefile
===================================================================
--- devel/scons/Makefile	(revision 466024)
+++ devel/scons/Makefile	(working copy)
@@ -3,6 +3,7 @@
 
 PORTNAME=	scons
 PORTVERSION=	3.0.1
+PORTREVISION=	1
 CATEGORIES=	devel python
 MASTER_SITES=	SF
 
@@ -14,7 +15,7 @@
 
 RUN_DEPENDS=	gm4:devel/m4
 
-USES=		python:2.7 shebangfix
+USES=		python:2.7+ shebangfix
 SHEBANG_FILES=	script/scons*
 USE_PYTHON=	distutils autoplist
 NO_ARCH=	yes
