Index: net/quagga/Makefile
===================================================================
--- net/quagga/Makefile	(revision 369706)
+++ net/quagga/Makefile	(working copy)
@@ -2,8 +2,7 @@
 # $FreeBSD$
 
 PORTNAME=	quagga
-PORTVERSION=	0.99.22.4
-PORTREVISION=	4
+PORTVERSION=	0.99.23.1
 CATEGORIES=	net ipv6
 MASTER_SITES=	SAVANNAH
 MASTER_SITE_SUBDIR=	quagga
Index: net/quagga/distinfo
===================================================================
--- net/quagga/distinfo	(revision 369706)
+++ net/quagga/distinfo	(working copy)
@@ -1,2 +1,2 @@
-SHA256 (quagga-0.99.22.4.tar.gz) = cbe48d5cc57bbaa07cfd8362ba598447dc94aa866ddc5794e57172709d36ba79
-SIZE (quagga-0.99.22.4.tar.gz) = 2352406
+SHA256 (quagga-0.99.23.1.tar.gz) = 3abf2046bc27539ce2d17c238e06c8fd0d479a8e402580c6aa455808bd48e004
+SIZE (quagga-0.99.23.1.tar.gz) = 2526612
Index: net/quagga/pkg-plist
===================================================================
--- net/quagga/pkg-plist	(revision 369706)
+++ net/quagga/pkg-plist	(working copy)
@@ -32,6 +32,7 @@
 include/quagga/if_rmap.h
 include/quagga/jhash.h
 include/quagga/keychain.h
+include/quagga/libospf.h
 include/quagga/linklist.h
 include/quagga/log.h
 include/quagga/md5.h
