Index: net/bird2/Makefile
===================================================================
--- net/bird2/Makefile	(revision 508373)
+++ net/bird2/Makefile	(working copy)
@@ -2,8 +2,7 @@
 # $FreeBSD$
 
 PORTNAME=	bird
-DISTVERSION=	2.0.4
-PORTREVISION=	2
+DISTVERSION=	2.0.5
 CATEGORIES=	net
 MASTER_SITES=	ftp://bird.network.cz/pub/bird/
 PKGNAMESUFFIX=	2
Index: net/bird2/distinfo
===================================================================
--- net/bird2/distinfo	(revision 508373)
+++ net/bird2/distinfo	(working copy)
@@ -1,3 +1,3 @@
-TIMESTAMP = 1553556348
-SHA256 (bird-2.0.4.tar.gz) = 676010b7517d4159b9af37401c26185f561ffcffeba73690a2ef2fad984714de
-SIZE (bird-2.0.4.tar.gz) = 1051911
+TIMESTAMP = 1565265942
+SHA256 (bird-2.0.5.tar.gz) = 4e4b736fd26579823a728be6a7746b3f525206e3c9a4a21fccb302cffd3029d3
+SIZE (bird-2.0.5.tar.gz) = 1099509
