Index: net/exabgp/Makefile
===================================================================
--- net/exabgp/Makefile	(revision 518433)
+++ net/exabgp/Makefile	(working copy)
@@ -2,45 +2,53 @@
 # $FreeBSD$
 
 PORTNAME=	exabgp
-PORTVERSION=	3.4.25
-PORTREVISION=	1
+PORTVERSION=	4.1.2
 CATEGORIES=	net
 MASTER_SITES=	GH \
 		ZI
+PKGNAMEPREFIX=	${PYTHON_PKGNAMEPREFIX}
 
 MAINTAINER=	zi@FreeBSD.org
 COMMENT=	BGP engine and route injector
 
-RUN_DEPENDS=	${PY_IPADDRESS}
-
 LICENSE=	BSD3CLAUSE
 LICENSE_FILE=	${WRKSRC}/COPYRIGHT
 
-USE_GITHUB=	yes
-GH_ACCOUNT=	Exa-Networks
-USES=		python:2.7
-SUB_FILES=	pkg-message exabgp.sh
+RUN_DEPENDS=	${PY_IPADDRESS}
+
+USES=		python shebangfix
+USE_PYTHON=	distutils autoplist
+
+SUB_FILES=	pkg-message
 SUB_LIST=	PYTHON_LIBDIR=${PYTHON_LIBDIR} PYTHON_CMD=${PYTHON_CMD} \
 		PYTHON_SITELIBDIR=${PYTHON_SITELIBDIR}
 USE_RC_SUBR=	${PORTNAME}
 NO_ARCH=	yes
 
+SHEBANG_FILES=	etc/exabgp/run/*.run \
+		etc/exabgp/run/*.pl \
+		etc/exabgp/run/*.py
+
+USE_GITHUB=	yes
+GH_ACCOUNT=	Exa-Networks
+
 USERS=		_bgpd
 GROUPS=		_bgpd
 
 OPTIONS_DEFINE=	EXAMPLES
 
-post-extract:
-	@${MV} -f ${WRKSRC}/etc/systemd ${WRKSRC}
+post-patch:
+	@${REINPLACE_CMD} -e 's|/etc/|${PREFIX}/etc/|' \
+	${WRKSRC}/doc/man/exabgp.1 \
+	${WRKSRC}/lib/exabgp/configuration/usage.py \
+	${WRKSRC}/lib/exabgp/configuration/process/parser.py \
+	${WRKSRC}/lib/exabgp/application/flow.py \
+	${WRKSRC}/lib/exabgp/application/healthcheck.py \
+	${WRKSRC}/lib/exabgp/application/bgp.py \
+ 	${WRKSRC}/lib/exabgp/application/cli.py
 
-do-build:
-	${PYTHON_CMD} -m compileall ${WRKSRC}/lib
-
-do-install:
-	${INSTALL_SCRIPT} ${WRKDIR}/${PORTNAME}.sh ${STAGEDIR}${PREFIX}/sbin/${PORTNAME}
-	${INSTALL_MAN} ${WRKSRC}/doc/man/${PORTNAME}.1 ${STAGEDIR}${PREFIX}/man/man1
+post-install:
 	@${MKDIR} ${STAGEDIR}${PYTHON_SITELIBDIR}/${PORTNAME} ${STAGEDIR}${ETCDIR}
-	(cd ${WRKSRC}/lib && ${COPYTREE_SHARE} . ${STAGEDIR}${PYTHON_SITELIBDIR})
 	${INSTALL_DATA} ${FILESDIR}/exabgp.env ${STAGEDIR}${ETCDIR}/exabgp.env.sample
 
 do-install-EXAMPLES-on:
Index: net/exabgp/distinfo
===================================================================
--- net/exabgp/distinfo	(revision 518433)
+++ net/exabgp/distinfo	(working copy)
@@ -1,3 +1,3 @@
-TIMESTAMP = 1531317755
-SHA256 (Exa-Networks-exabgp-3.4.25_GH0.tar.gz) = ea09c2ba0babda836887665e2de2973263f5f06cb6dc80d253fbc0e94f729984
-SIZE (Exa-Networks-exabgp-3.4.25_GH0.tar.gz) = 544345
+TIMESTAMP = 1565637826
+SHA256 (Exa-Networks-exabgp-4.1.2_GH0.tar.gz) = 5921e002f196e814d02349a15c250b1fc8bf45c7299b6b652d2fed04eebb529a
+SIZE (Exa-Networks-exabgp-4.1.2_GH0.tar.gz) = 2912897
Index: net/exabgp/files/exabgp.in
===================================================================
--- net/exabgp/files/exabgp.in	(revision 518433)
+++ net/exabgp/files/exabgp.in	(working copy)
@@ -25,7 +25,7 @@
 required_files=${exabgp_conf}
 
 command="%%PYTHON_CMD%%"
-command_args="-m exabgp --folder /usr/local/etc/exabgp ${exabgp_conf}"
+command_args="-m exabgp ${exabgp_conf}"
 
 configtest_cmd="exabgp_configtest"
 start_precmd="exabgp_precmd"
@@ -45,7 +45,7 @@
 exabgp_configtest()
 {
 	echo "Performing sanity check on ${name} configuration."
-	result=$(env exabgp.log.configuration=true exabgp.daemon.daemonize=false exabgp.log.level=ERR exabgp.log.destination=stdout ${command} -m ${name} --folder /usr/local/etc/exabgp -t ${exabgp_conf} 2>&1)
+	result=$(env exabgp.log.configuration=true exabgp.daemon.daemonize=false exabgp.log.level=ERR exabgp.log.destination=stdout ${command} -m ${name} /usr/local/etc/exabgp -t ${exabgp_conf} 2>&1)
 	if [ ! -z "${result}" ]; then
 		echo "${result}"
 		err 1 "Failed to validate configuration file, aborting."
@@ -54,9 +54,14 @@
 
 exabgp_precmd()
 {
+	if ! [ -d /var/run/${name} ]; then
+		install -d -o _bgpd -g _bgpd -m 755 /var/run/${name}
+		install -d -o _bgpd -g _bgpd -m 755 /var/log/${name}
+		mkfifo -m 600 /var/run/exabgp/exabgp.in
+		mkfifo -m 600 /var/run/exabgp/exabgp.out
+		chown _bgpd:_bgpd /var/run/exabgp/exabgp.*
+	fi
 	exabgp_configtest
-	install -d -o _bgpd -g _bgpd -m 755 /var/run/${name}
-	install -d -o _bgpd -g _bgpd -m 755 /var/log/${name}
 	rm -f ${pidfile}
 }
 
Index: net/exabgp/files/patch-setup.py
===================================================================
--- net/exabgp/files/patch-setup.py	(nonexistent)
+++ net/exabgp/files/patch-setup.py	(working copy)
@@ -0,0 +1,10 @@
+--- setup.py.orig	2018-03-30 08:54:47 UTC
++++ setup.py
+@@ -348,7 +348,6 @@ os_name = platform.system()
+ 
+ files_definition = [
+ 	('share/exabgp/processes',filesOf('etc/exabgp')),
+-	('share/exabgp/etc',testFilesOf('qa/conf')),
+ ]
+ 
+ if os_name != 'NetBSD':

Property changes on: net/exabgp/files/patch-setup.py
___________________________________________________________________
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Index: net/exabgp/pkg-plist
===================================================================
--- net/exabgp/pkg-plist	(revision 518433)
+++ net/exabgp/pkg-plist	(working copy)
@@ -1,391 +1,97 @@
 @sample etc/exabgp/exabgp.env.sample
-man/man1/exabgp.1.gz
-sbin/exabgp
-%%PYTHON_SITELIBDIR%%/exabgp/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/__main__.py
-%%PYTHON_SITELIBDIR%%/exabgp/__main__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/application/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/application/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/application/bgp.py
-%%PYTHON_SITELIBDIR%%/exabgp/application/bgp.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/application/bmp.py
-%%PYTHON_SITELIBDIR%%/exabgp/application/bmp.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/application/cli.py
-%%PYTHON_SITELIBDIR%%/exabgp/application/cli.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/application/healthcheck.py
-%%PYTHON_SITELIBDIR%%/exabgp/application/healthcheck.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/application/netlink.py
-%%PYTHON_SITELIBDIR%%/exabgp/application/netlink.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/keepalive.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/keepalive.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/message.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/message.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/nop.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/nop.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/notification.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/notification.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/asn.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/asn.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/addpath.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/addpath.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/asn4.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/asn4.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/capabilities.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/capabilities.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/capability.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/capability.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/graceful.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/graceful.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/mp.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/mp.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/ms.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/ms.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/negotiated.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/negotiated.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/operational.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/operational.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/refresh.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/refresh.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/unknown.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/capability/unknown.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/holdtime.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/holdtime.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/routerid.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/routerid.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/version.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/open/version.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/operational.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/operational.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/refresh.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/refresh.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/unknown.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/unknown.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/aggregator.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/aggregator.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/aigp.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/aigp.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/aspath.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/aspath.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/atomicaggregate.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/atomicaggregate.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/attribute.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/attribute.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/attributes.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/attributes.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/clusterlist.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/clusterlist.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/communities.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/communities.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/community.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/community.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/communities.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/communities.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/community.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/community.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/encapsulation.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/encapsulation.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/l2info.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/l2info.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/origin.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/origin.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/rt.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/rt.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/traffic.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/community/extended/traffic.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/generic.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/generic.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/localpref.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/localpref.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/med.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/med.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/mprnlri.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/mprnlri.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/mpurnlri.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/mpurnlri.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/nexthop.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/nexthop.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/origin.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/origin.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/originatorid.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/originatorid.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/pmsi.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/attribute/pmsi.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/eor.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/eor.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/cidr.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/cidr.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/evpn/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/evpn/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/evpn/ethernetad.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/evpn/ethernetad.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/evpn/mac.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/evpn/mac.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/evpn/multicast.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/evpn/multicast.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/evpn/nlri.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/evpn/nlri.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/evpn/segment.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/evpn/segment.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/flow.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/flow.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/mpls.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/mpls.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/nlri.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/nlri.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/prefix.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/prefix.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/esi.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/esi.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/etag.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/etag.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/labels.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/labels.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/mac.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/mac.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/path.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/path.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/rd.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/qualifier/rd.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/rtc.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/rtc.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/vpls.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/message/update/nlri/vpls.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/neighbor.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/neighbor.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/timer.py
-%%PYTHON_SITELIBDIR%%/exabgp/bgp/timer.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bmp/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/bmp/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bmp/header.py
-%%PYTHON_SITELIBDIR%%/exabgp/bmp/header.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bmp/message.py
-%%PYTHON_SITELIBDIR%%/exabgp/bmp/message.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bmp/negotiated.py
-%%PYTHON_SITELIBDIR%%/exabgp/bmp/negotiated.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/bmp/peer.py
-%%PYTHON_SITELIBDIR%%/exabgp/bmp/peer.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/ancient.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/ancient.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/asn.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/asn.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/capability.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/capability.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/family.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/family.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/neighbor.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/neighbor.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/process.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/process.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/session.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bgp/session.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bmp/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/bmp/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/check.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/check.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/location.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/location.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/parser.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/parser.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/raised.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/raised.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/reader.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/reader.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/section.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/section.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/tokeniser.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/engine/tokeniser.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/environment.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/environment.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/format.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/format.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/neighbor/capability.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/neighbor/capability.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/setup.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/setup.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/show/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/show/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/usage.py
-%%PYTHON_SITELIBDIR%%/exabgp/configuration/usage.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/data/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/data/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/data/check.py
-%%PYTHON_SITELIBDIR%%/exabgp/data/check.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/debug.py
-%%PYTHON_SITELIBDIR%%/exabgp/debug.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/dep/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/dep/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/dep/cmd2.py
-%%PYTHON_SITELIBDIR%%/exabgp/dep/cmd2.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/dep/counter.py
-%%PYTHON_SITELIBDIR%%/exabgp/dep/counter.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/dep/docopt.py
-%%PYTHON_SITELIBDIR%%/exabgp/dep/docopt.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/dep/gcdump.py
-%%PYTHON_SITELIBDIR%%/exabgp/dep/gcdump.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/dep/ipaddress.py
-%%PYTHON_SITELIBDIR%%/exabgp/dep/ipaddress.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/dep/lsprofcalltree.py
-%%PYTHON_SITELIBDIR%%/exabgp/dep/lsprofcalltree.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/dep/objgraph.py
-%%PYTHON_SITELIBDIR%%/exabgp/dep/objgraph.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/dep/ordereddict.py
-%%PYTHON_SITELIBDIR%%/exabgp/dep/ordereddict.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/dep/profiler.py
-%%PYTHON_SITELIBDIR%%/exabgp/dep/profiler.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/dep/pyparsing.py
-%%PYTHON_SITELIBDIR%%/exabgp/dep/pyparsing.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/logger.py
-%%PYTHON_SITELIBDIR%%/exabgp/logger.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/attributes.py
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/attributes.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/firewall.py
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/firewall.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/message.py
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/message.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/old.py
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/old.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/route/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/route/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/route/address.py
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/route/address.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/route/link.py
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/route/link.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/route/neighbor.py
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/route/neighbor.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/route/network.py
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/route/network.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/sequence.py
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/sequence.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/tc.py
-%%PYTHON_SITELIBDIR%%/exabgp/netlink/tc.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/enum.py
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/enum.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/family.py
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/family.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/ip/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/ip/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/ip/address.py
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/ip/address.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/ip/fragment.py
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/ip/fragment.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/ip/icmp.py
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/ip/icmp.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/ip/tcp/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/ip/tcp/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/ip/tcp/flag.py
-%%PYTHON_SITELIBDIR%%/exabgp/protocol/ip/tcp/flag.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/command.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/command.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/control.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/control.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/decoder/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/decoder/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/decoder/json.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/decoder/json.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/decoder/text.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/decoder/text.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/encoding.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/encoding.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/processes.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/api/processes.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/daemon.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/daemon.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/listener.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/listener.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/loop.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/loop.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/network/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/network/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/network/connection.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/network/connection.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/network/error.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/network/error.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/network/incoming.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/network/incoming.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/network/outgoing.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/network/outgoing.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/network/tcp.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/network/tcp.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/peer.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/peer.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/protocol.py
-%%PYTHON_SITELIBDIR%%/exabgp/reactor/protocol.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/rib/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/rib/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/rib/change.py
-%%PYTHON_SITELIBDIR%%/exabgp/rib/change.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/rib/store.py
-%%PYTHON_SITELIBDIR%%/exabgp/rib/store.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/util/__init__.py
-%%PYTHON_SITELIBDIR%%/exabgp/util/__init__.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/util/cache.py
-%%PYTHON_SITELIBDIR%%/exabgp/util/cache.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/util/coroutine.py
-%%PYTHON_SITELIBDIR%%/exabgp/util/coroutine.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/util/dictionary.py
-%%PYTHON_SITELIBDIR%%/exabgp/util/dictionary.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/util/enumeration.py
-%%PYTHON_SITELIBDIR%%/exabgp/util/enumeration.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/util/errstr.py
-%%PYTHON_SITELIBDIR%%/exabgp/util/errstr.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/util/hashtable.py
-%%PYTHON_SITELIBDIR%%/exabgp/util/hashtable.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/util/ip.py
-%%PYTHON_SITELIBDIR%%/exabgp/util/ip.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/util/od.py
-%%PYTHON_SITELIBDIR%%/exabgp/util/od.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/util/panic.py
-%%PYTHON_SITELIBDIR%%/exabgp/util/panic.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/util/trace.py
-%%PYTHON_SITELIBDIR%%/exabgp/util/trace.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/util/usage.py
-%%PYTHON_SITELIBDIR%%/exabgp/util/usage.pyc
-%%PYTHON_SITELIBDIR%%/exabgp/version.py
-%%PYTHON_SITELIBDIR%%/exabgp/version.pyc
-%%PORTEXAMPLES%%%%EXAMPLESDIR%%/ibgp4-healthcheck.txt
-%%PORTEXAMPLES%%%%EXAMPLESDIR%%/ibgp4-tcpcontrol.txt
-%%PORTEXAMPLES%%%%EXAMPLESDIR%%/processes/dynamic-1.pl
-%%PORTEXAMPLES%%%%EXAMPLESDIR%%/processes/dynamic-1.sh
-%%PORTEXAMPLES%%%%EXAMPLESDIR%%/processes/read-write.py
-%%PORTEXAMPLES%%%%EXAMPLESDIR%%/processes/stderr-1.py
-%%PORTEXAMPLES%%%%EXAMPLESDIR%%/processes/syslog-1.py
-%%PORTEXAMPLES%%%%EXAMPLESDIR%%/processes/tcp-server
-%%PORTEXAMPLES%%%%EXAMPLESDIR%%/processes/watchdog-1.pl
-%%PORTEXAMPLES%%%%EXAMPLESDIR%%/processes/watchdog-1.sh
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-add-remove.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-announce.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-announcement.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-api.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-attributes.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-attributes-vpn.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-broken-flow.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-check.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-eor.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-fast.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-flow.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-manual-eor.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-multi-neighbor.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-multisession.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-nexthop-self.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-nexthop.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-notification.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-open.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-reload.1.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-reload.2.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-teardown.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-vpls.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-vpnv4.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/api-watchdog.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-addpath.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-aggregator.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-attributes.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-flow-redirect.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-flow.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-generic-attribute.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-group-limit.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-group.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-ipself4.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-ipself6.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-ipv46routes4family.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-ipv46routes6family.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-ipv6grouping.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-l2vpn.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-largecommunity.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-name.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-new-v4.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-new-v6.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-no-asn4.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-parity.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-path-information.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-prefix-sid.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-split.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-template.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-unknowncap.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/conf-vpn.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/example-api-program.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/example-healthcheck.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/example-tcp-control.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/example-tcp-control.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/extended-nexthop.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/parse-community.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/parse-dual-neighbor.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/parse-md5.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/parse-multiple-process.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/parse-multisession.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/parse-process.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/parse-simple-v4.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/parse-simple-v6.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/parse-ttl.conf
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-add-remove.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-announce.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-announcement.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-api.nothing.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-api.receive.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-attributes.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-attributes-vpn.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-broken-flow.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-check.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-eor.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-fast.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-flow.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-manual-eor.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-multi-neighbor.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-multisession.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-nexthop-self.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-nexthop.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-notification.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-open.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-reload.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-teardown.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-vpls.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/api-vpnv4.run
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/dynamic-1.pl
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/dynamic-1.sh
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/stderr-1.py
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/syslog-1.py
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/watchdog-1.pl
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/watchdog-1.sh
+%%PORTEXAMPLES%%%%EXAMPLESDIR%%/run/watchdog.run
 @dir etc/exabgp
-@unexec if [ -d %D/%%ETCDIR%% ]; then echo "==> If you are permanently removing this port, you should do a ``rm -rf ${PKG_PREFIX}/etc/exabgp`` to remove any configuration files left."; fi
+@postunexec if [ -d %D/%%ETCDIR%% ]; then echo "==> If you are permanently removing this port, you should do a ``rm -rf ${PKG_PREFIX}/etc/exabgp`` to remove any configuration files left."; fi
