Index: net/frr/files/frr.in
===================================================================
--- net/frr/files/frr.in	(revision 444990)
+++ net/frr/files/frr.in	(working copy)
@@ -5,7 +5,7 @@
 
 # PROVIDE: frr
 # REQUIRE: netif routing 
-# KEYWORD: nojail shutdown
+# KEYWORD: nojailvnet shutdown
 
 # Add the following line to /etc/rc.conf to enable frr:
 #  frr_enable="YES"
@@ -78,7 +78,7 @@
 		echo Waiting for ${frr_wait_for} route...
 		while [ ${waited_for} -lt ${frr_wait_seconds} ]; do
 			/sbin/route -n get ${frr_wait_for} >/dev/null 2>&1 && break;
-			waited_for=$((${waited_for}+1))
+			waited_for=$((waited_for+1))
 			sleep 1;
 		done
 		[ ${waited_for} -lt ${frr_wait_seconds} ] || echo Giving up...
@@ -93,13 +93,13 @@
 	    command=%%PREFIX%%/sbin/${daemon}
 	    required_files=%%ETCDIR%%/${daemon}.conf
 	    pidfile=/var/run/frr/${daemon}.pid
-	    if [ ${frr_cmd} = "restart" -o ${frr_cmd} = "start" ]; then
+	    if [ ${frr_cmd} = "restart" ] || [ ${frr_cmd} = "start" ]; then
 		check_config
 	    fi
-	    if [ ${frr_cmd} = "start" -a ! -f ${required_files} ]; then
+	    if [ ${frr_cmd} = "start" ] && ![ -f ${required_files} ]; then
 			continue
 	    fi
-	    if [ ${frr_cmd} = "stop" -a -z $(check_process ${command}) ]; then
+	    if [ ${frr_cmd} = "stop" ] && [ -z $(check_process ${command}) ]; then
 			continue
 	    fi
 	    eval flags=\$\{${daemon}_flags:-\"${frr_flags}\"\}
@@ -123,6 +123,9 @@
     fast*)
 	frr_cmd=${frr_cmd#fast}
 	;;
+	quiet*)
+	frr_cmd=${frr_cmd#quiet}
+	;;
 esac
 shift
 
Index: net/frr/files/watchfrr.in
===================================================================
--- net/frr/files/watchfrr.in	(revision 444990)
+++ net/frr/files/watchfrr.in	(working copy)
@@ -5,13 +5,13 @@
 
 # PROVIDE: watchfrr
 # REQUIRE: NETWORKING frr
-# KEYWORD: shutdown
+# KEYWORD: nojailvtnet shutdown
 
 # Add the following line to /etc/rc.conf to enable watchfrr:
 # watchfrr_enable="YES"
 #
 # You may also wish to use the following variables to fine-tune startup:
-# watchquagga_flags
+# watchfrr_flags
 
 . /etc/rc.subr
 
