diff --git sysutils/dtrace-toolkit/Makefile sysutils/dtrace-toolkit/Makefile
index f741b6fb7434..fe2ba46cfcb0 100644
--- sysutils/dtrace-toolkit/Makefile
+++ sysutils/dtrace-toolkit/Makefile
@@ -10,7 +10,7 @@ WWW=		https://github.com/opendtrace/toolkit
 LICENSE=	CDDL
 LICENSE_FILE=	${WRKSRC}/Docs/cddl1.txt
 
-USES=		perl5 python ruby shebangfix
+USES=		perl5 python shebangfix
 USE_GITHUB=	yes
 GH_ACCOUNT=	opendtrace
 GH_PROJECT=	toolkit
