diff --git sysutils/mstflint/Makefile sysutils/mstflint/Makefile
index 196b63f1e3b3..5002d784ab41 100644
--- sysutils/mstflint/Makefile
+++ sysutils/mstflint/Makefile
@@ -44,7 +44,6 @@ USES+=	ssl
 .if ${PORT_OPTIONS:MADAB}
 CONFIGURE_ARGS+=	--enable-adb-generic-tools
 .if ${PORT_OPTIONS:MSTATIC}
-BROKEN=		does not link: expat2 no longer provides static library
 BUILD_DEPENDS+=	xmlwf:textproc/expat2 \
 		boost-libs>0:devel/boost-libs
 .else
