Index: sbin/dhclient/dhclient-script
===================================================================
--- sbin/dhclient/dhclient-script	(revision 276174)
+++ sbin/dhclient/dhclient-script	(working copy)
@@ -175,6 +175,8 @@
 			else
 				route add default $router >/dev/null 2>&1
 			fi
+			# Set LED 2 ON if default route learned
+			[ -e /dev/led/led2 ] && echo "1" > /dev/led/led2
 		fi
 		# 2nd and subsequent default routers error out, so explicitly
 		# stop processing the list after the first one.
@@ -319,6 +321,9 @@
 	;;
 
 PREINIT)
+    # Start to blink LED 2 slowly when looking for IP address
+    [ -e /dev/led/led2 ] && echo f9 > /dev/led/led2
+
 	delete_old_alias
 	$IFCONFIG $interface inet alias 0.0.0.0 netmask 255.0.0.0 broadcast 255.255.255.255 up
 	;;
