Index: security/snort/Makefile
===================================================================
--- security/snort/Makefile	(revision 405672)
+++ security/snort/Makefile	(working copy)
@@ -101,7 +101,7 @@
 LIBNET_LIBDIR=	${LIBNET_LIBS:M-L*:S/-L//}
 
 CFLAGS+=	-fstack-protector
-CONFIGURE_ARGS+=--enable-reload \
+CONFIGURE_ARGS+=--enable-reload --enable-pthread \
 		--enable-mpls --enable-targetbased \
 		--enable-reload-error-restart \
 		--with-dnet-includes=${LIBNET_INCDIR} \
