Index: sys/dev/oce/oce_mbox.c
===================================================================
--- sys/dev/oce/oce_mbox.c	(revision 284539)
+++ sys/dev/oce/oce_mbox.c	(working copy)
@@ -865,7 +865,7 @@
 		req->iface_flags = MBX_RX_IFACE_FLAGS_PROMISCUOUS;
 
 	if (enable & 0x02)
-		req->iface_flags = MBX_RX_IFACE_FLAGS_VLAN_PROMISCUOUS;
+		req->iface_flags |= MBX_RX_IFACE_FLAGS_VLAN_PROMISCUOUS;
 
 	req->if_id = sc->if_id;
 
