Index: net/dhcprelya/Makefile
===================================================================
--- net/dhcprelya/Makefile	(revision 356884)
+++ net/dhcprelya/Makefile	(working copy)
@@ -2,18 +2,18 @@
 # $FreeBSD$
 
 PORTNAME=	dhcprelya
-PORTVERSION=	4.7
+PORTVERSION=	4.8
 CATEGORIES=	net
-MASTER_SITES=	GH
 
 MAINTAINER=	sem@FreeBSD.org
 COMMENT=	Lightweight DHCP relay agent. Yandex edition
 
-LICENSE=	BSD
+LICENSE=	BSD4CLAUSE
 
 USE_GITHUB=	yes
 GH_ACCOUNT=	sem-hub
-GH_COMMIT=	6df4e97
+GH_TAGNAME=	${GH_COMMIT}
+GH_COMMIT=	32d3356
 
 USE_RC_SUBR=	dhcprelya
 
@@ -26,8 +26,6 @@
 #OPTIONS_DEFINE=	RADIUS_PLUGIN
 RADIUS_PLUGIN_PLUGIN_DESC=	build radius plugin
 
-NO_STAGE=	yes
-
 OPTIONS_DEFINE=	DOCS
 
 .include <bsd.port.options.mk>
@@ -38,12 +36,12 @@
 .endif
 
 post-patch:
-	@${REINPLACE_CMD} -e 's#/usr/local#${PREFIX}#' ${WRKSRC}/Makefile
+	@${REINPLACE_CMD} -e 's#/usr/local#${STAGEDIR}/${PREFIX}#' ${WRKSRC}/Makefile
 
 post-install:
 .if ${PORT_OPTIONS:MDOCS}
-	${MKDIR} ${DOCSDIR}
-	${INSTALL_DATA} ${WRKSRC}/README ${DOCSDIR}
+	${MKDIR} ${STAGEDIR}/${DOCSDIR}
+	${INSTALL_DATA} ${WRKSRC}/README ${STAGEDIR}/${DOCSDIR}
 .endif
 
 .include <bsd.port.mk>
Index: net/dhcprelya/distinfo
===================================================================
--- net/dhcprelya/distinfo	(revision 356884)
+++ net/dhcprelya/distinfo	(working copy)
@@ -1,2 +1,2 @@
-SHA256 (dhcprelya-4.7.tar.gz) = 0b59ec0276d89ec7be50aadad4d82acea8b154c877a822f69971c25725a660cd
-SIZE (dhcprelya-4.7.tar.gz) = 24192
+SHA256 (dhcprelya-4.8.tar.gz) = 4a224a6f2ba35df54e9c2d4b399c7b9b279522dd6ac9c65d3dae4bfba7a5fc9c
+SIZE (dhcprelya-4.8.tar.gz) = 24212
