Index: security/ipsec-tools/Makefile
===================================================================
--- security/ipsec-tools/Makefile	(revision 307785)
+++ security/ipsec-tools/Makefile	(working copy)
@@ -30,6 +30,7 @@
 CONFIGURE_ARGS=	--enable-shared --sysconfdir=${PREFIX}/etc/racoon \
 		--localstatedir=${STATEDIR:S/\/racoon//} \
 		--with-pkgversion=freebsd-${PORTVERSION}
+USE_GCC=	4.2+
 
 STATEDIR=	/var/db/racoon
 SUB_LIST+=	STATEDIR=${STATEDIR}
