Index: net/quagga/files/patch-lib-zebra.h
===================================================================
--- files/patch-lib-zebra.h	(revision 0)
+++ files/patch-lib-zebra.h	(working copy)
@@ -0,0 +1,12 @@
+--- lib/zebra.h.orig	2014-03-16 09:54:05.510349508 +0100
++++ lib/zebra.h	2014-03-16 10:57:36.640087484 +0100
+@@ -355,7 +355,8 @@
+  * OpenBSD: network byte order, apart from older versions which are as per 
+  *          *BSD
+  */
+-#if defined(__NetBSD__) || defined(__FreeBSD__) \
++#if defined(__NetBSD__) \
++   || (defined(__FreeBSD__) && (__FreeBSD_version < 1100030)) \
+    || (defined(__OpenBSD__) && (OpenBSD < 200311)) \
+    || (defined(SUNOS_5) && defined(WORDS_BIGENDIAN))
+ #define HAVE_IP_HDRINCL_BSD_ORDER

Property changes on: net/quagga/files/patch-lib-zebra.h
___________________________________________________________________
Added: svn:mime-type
## -0,0 +1 ##
+text/plain
\ No newline at end of property
Added: fbsd:nokeywords
## -0,0 +1 ##
+yes
\ No newline at end of property
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Index: net/quagga/Makefile
===================================================================
--- net/quagga/Makefile	(revision 366911)
+++ net/quagga/Makefile	(working copy)
@@ -3,7 +3,7 @@
 
 PORTNAME=	quagga
 PORTVERSION=	0.99.22.4
-PORTREVISION=	3
+PORTREVISION=	4
 CATEGORIES=	net ipv6
 MASTER_SITES=	SAVANNAH
 MASTER_SITE_SUBDIR=	quagga
